

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Genotype &mdash; Hail</title>
  

  
  
    <link rel="shortcut icon" href="../hail_logo_sq.png"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="../navbar.css" type="text/css" />
  
    <link rel="stylesheet" href="../_static/rtd_modifications.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="../genindex.html"/>
        <link rel="search" title="Search" href="../search.html"/>
    <link rel="top" title="Hail" href="../index.html"/>
        <link rel="up" title="representation" href="index.html"/>
        <link rel="next" title="Call" href="hail.representation.Call.html"/>
        <link rel="prev" title="AltAllele" href="hail.representation.AltAllele.html"/> 

  
  <script src="../_static/js/modernizr.min.js"></script>
  
  

</head>

<body class="wy-body-for-nav" role="document">

  <nav class="navbar navbar-default navbar-static-top" id="hail-navbar">
<div class="container-fluid" id="hail-container-fluid">
    <div class="navbar-header" id="hail-navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#hail-navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
   
      <a class="navbar-left" id="hail-navbar-brand" href="../index.html"><img alt="Hail" id="logo" src="hail-logo-cropped.png"></a>

    </div>

    <div class="collapse navbar-collapse" id="hail-navbar-collapse">
      <ul class="nav navbar-nav navbar-right">
          <li id="home" class="nav-item"><a href="../index.html">HOME</a></li>
          <li id="docs" class="nav-item"><a href="index.html">DOCS</a></li>
          <li id="forum" class="nav-item"><a href="http://discuss.hail.is">FORUM</a></li>
          <li id="chat" class="nav-item dropdown">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">CHAT<span class="caret"></span></a>
          <ul class="dropdown-menu dropdown-menu-left">
            <li><a class="dropdown-item" href="https://gitter.im/hail-is/hail">General</a></li>
            <li><a class="dropdown-item" href="https://gitter.im/hail-is/hail-dev">Developers</a></li>
          </ul>
        </li>
       <li id="code" class="nav-item"><a href="https://github.com/hail-is/hail">CODE</a></li>
      </ul>
    </div>
  </div>
</nav>

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="../index.html" class="icon icon-home"> Hail
          

          
          </a>

          
            
            
              <div class="version">
                devel
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="../exprlang.html">Expression Language</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../hail.HailContext.html">HailContext</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hail.VariantDataset.html">VariantDataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hail.KeyTable.html">KeyTable</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hail.TextTableConfig.html">TextTableConfig</a></li>
<li class="toctree-l2"><a class="reference internal" href="../hail.KinshipMatrix.html">KinshipMatrix</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="index.html">representation</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="hail.representation.Variant.html">Variant</a></li>
<li class="toctree-l3"><a class="reference internal" href="hail.representation.AltAllele.html">AltAllele</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Genotype</a></li>
<li class="toctree-l3"><a class="reference internal" href="hail.representation.Call.html">Call</a></li>
<li class="toctree-l3"><a class="reference internal" href="hail.representation.Locus.html">Locus</a></li>
<li class="toctree-l3"><a class="reference internal" href="hail.representation.Interval.html">Interval</a></li>
<li class="toctree-l3"><a class="reference internal" href="hail.representation.IntervalTree.html">IntervalTree</a></li>
<li class="toctree-l3"><a class="reference internal" href="hail.representation.Struct.html">Struct</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../expr/index.html">expr</a></li>
<li class="toctree-l2"><a class="reference internal" href="../utils/index.html">utils</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../annotationdb.html">Annotation Database</a></li>
<li class="toctree-l1"><a class="reference internal" href="../other_resources.html">Other Resources</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">Hail</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../api.html">Python API</a> &raquo;</li>
        
          <li><a href="index.html">representation</a> &raquo;</li>
        
      <li>Genotype</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../_sources/representation/hail.representation.Genotype.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

/* nice headers on first paragraph of info/warning boxes */
.admonition .first {
    margin: -12px;
    padding: 6px 12px;
    margin-bottom: 12px;
    color: #fff;
    line-height: 1;
    display: block;
}
.admonition.warning .first {
    background: #f0b37e;
}
.admonition.note .first {
    background: #6ab0de;
}
.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="genotype">
<h1>Genotype<a class="headerlink" href="#genotype" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="hail.representation.Genotype">
<em class="property">class </em><code class="descclassname">hail.representation.</code><code class="descname">Genotype</code><span class="sig-paren">(</span><em>gt</em>, <em>ad=None</em>, <em>dp=None</em>, <em>gq=None</em>, <em>pl=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype" title="Permalink to this definition">¶</a></dt>
<dd><p>An object that represents an individual&#8217;s genotype at a genomic locus.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gt</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; Genotype hard call</li>
<li><strong>ad</strong> (<em>list of int</em><em> or </em><em>None</em>) &#8211; allelic depth (1 element per allele including reference)</li>
<li><strong>dp</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; total depth</li>
<li><strong>gq</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; genotype quality</li>
<li><strong>pl</strong> (<em>list of int</em><em> or </em><em>None</em>) &#8211; phred-scaled posterior genotype likelihoods (1 element per possible genotype)</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.ad" title="hail.representation.Genotype.ad"><code class="xref py py-obj docutils literal"><span class="pre">ad</span></code></a></td>
<td>Returns the allelic depth.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.dp" title="hail.representation.Genotype.dp"><code class="xref py py-obj docutils literal"><span class="pre">dp</span></code></a></td>
<td>Returns the total depth.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.gq" title="hail.representation.Genotype.gq"><code class="xref py py-obj docutils literal"><span class="pre">gq</span></code></a></td>
<td>Returns the phred-scaled genotype quality.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.gt" title="hail.representation.Genotype.gt"><code class="xref py py-obj docutils literal"><span class="pre">gt</span></code></a></td>
<td>Returns the hard genotype call.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.pl" title="hail.representation.Genotype.pl"><code class="xref py py-obj docutils literal"><span class="pre">pl</span></code></a></td>
<td>Returns the phred-scaled genotype posterior likelihoods.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">__init__</span></code></td>
<td>Initialize a Genotype object.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.dosage" title="hail.representation.Genotype.dosage"><code class="xref py py-obj docutils literal"><span class="pre">dosage</span></code></a></td>
<td>Returns the linear-scaled genotype probabilities.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.fraction_reads_ref" title="hail.representation.Genotype.fraction_reads_ref"><code class="xref py py-obj docutils literal"><span class="pre">fraction_reads_ref</span></code></a></td>
<td>Returns the fraction of reads that are reference reads.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.is_called" title="hail.representation.Genotype.is_called"><code class="xref py py-obj docutils literal"><span class="pre">is_called</span></code></a></td>
<td>True if the genotype call is non-missing.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.is_called_non_ref" title="hail.representation.Genotype.is_called_non_ref"><code class="xref py py-obj docutils literal"><span class="pre">is_called_non_ref</span></code></a></td>
<td>True if the genotype call contains any non-reference alleles.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.is_het" title="hail.representation.Genotype.is_het"><code class="xref py py-obj docutils literal"><span class="pre">is_het</span></code></a></td>
<td>True if the genotype call contains two different alleles.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.is_het_non_ref" title="hail.representation.Genotype.is_het_non_ref"><code class="xref py py-obj docutils literal"><span class="pre">is_het_non_ref</span></code></a></td>
<td>True if the genotype call contains two different alternate alleles.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.is_het_ref" title="hail.representation.Genotype.is_het_ref"><code class="xref py py-obj docutils literal"><span class="pre">is_het_ref</span></code></a></td>
<td>True if the genotype call contains one reference and one alternate allele.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.is_hom_ref" title="hail.representation.Genotype.is_hom_ref"><code class="xref py py-obj docutils literal"><span class="pre">is_hom_ref</span></code></a></td>
<td>True if the genotype call is 0/0</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.is_hom_var" title="hail.representation.Genotype.is_hom_var"><code class="xref py py-obj docutils literal"><span class="pre">is_hom_var</span></code></a></td>
<td>True if the genotype call contains two identical alternate alleles.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.is_not_called" title="hail.representation.Genotype.is_not_called"><code class="xref py py-obj docutils literal"><span class="pre">is_not_called</span></code></a></td>
<td>True if the genotype call is missing.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.num_alt_alleles" title="hail.representation.Genotype.num_alt_alleles"><code class="xref py py-obj docutils literal"><span class="pre">num_alt_alleles</span></code></a></td>
<td>Returns the count of non-reference alleles.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.od" title="hail.representation.Genotype.od"><code class="xref py py-obj docutils literal"><span class="pre">od</span></code></a></td>
<td>Returns the difference between the total depth and the allelic depth sum.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.one_hot_alleles" title="hail.representation.Genotype.one_hot_alleles"><code class="xref py py-obj docutils literal"><span class="pre">one_hot_alleles</span></code></a></td>
<td>Returns a list containing the one-hot encoded representation of the called alleles.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.representation.Genotype.one_hot_genotype" title="hail.representation.Genotype.one_hot_genotype"><code class="xref py py-obj docutils literal"><span class="pre">one_hot_genotype</span></code></a></td>
<td>Returns a list containing the one-hot encoded representation of the genotype call.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.representation.Genotype.p_ab" title="hail.representation.Genotype.p_ab"><code class="xref py py-obj docutils literal"><span class="pre">p_ab</span></code></a></td>
<td>Returns the p-value associated with finding the given allele depth ratio.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="hail.representation.Genotype.ad">
<code class="descname">ad</code><a class="headerlink" href="#hail.representation.Genotype.ad" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the allelic depth.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.dosage">
<code class="descname">dosage</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.dosage"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.dosage" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the linear-scaled genotype probabilities.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of float</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="hail.representation.Genotype.dp">
<code class="descname">dp</code><a class="headerlink" href="#hail.representation.Genotype.dp" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the total depth.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.fraction_reads_ref">
<code class="descname">fraction_reads_ref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.fraction_reads_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.fraction_reads_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the fraction of reads that are reference reads.</p>
<p>Equivalent to:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">g</span><span class="o">.</span><span class="n">ad</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">/</span> <span class="nb">sum</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">ad</span><span class="p">)</span>
</pre></div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">float or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="hail.representation.Genotype.gq">
<code class="descname">gq</code><a class="headerlink" href="#hail.representation.Genotype.gq" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the phred-scaled genotype quality.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="hail.representation.Genotype.gt">
<code class="descname">gt</code><a class="headerlink" href="#hail.representation.Genotype.gt" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the hard genotype call.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_called">
<code class="descname">is_called</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_called"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_called" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call is non-missing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_called_non_ref">
<code class="descname">is_called_non_ref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_called_non_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_called_non_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call contains any non-reference alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_het">
<code class="descname">is_het</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_het"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_het" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call contains two different alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_het_non_ref">
<code class="descname">is_het_non_ref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_het_non_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_het_non_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call contains two different alternate alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_het_ref">
<code class="descname">is_het_ref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_het_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_het_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call contains one reference and one alternate allele.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_hom_ref">
<code class="descname">is_hom_ref</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_hom_ref"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_hom_ref" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call is 0/0</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_hom_var">
<code class="descname">is_hom_var</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_hom_var"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_hom_var" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call contains two identical alternate alleles.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.is_not_called">
<code class="descname">is_not_called</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.is_not_called"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.is_not_called" title="Permalink to this definition">¶</a></dt>
<dd><p>True if the genotype call is missing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">bool</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.num_alt_alleles">
<code class="descname">num_alt_alleles</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.num_alt_alleles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.num_alt_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the count of non-reference alleles.</p>
<p>This function returns None if the genotype call is missing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.od">
<code class="descname">od</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.od"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.od" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the difference between the total depth and the allelic depth sum.</p>
<p>Equivalent to:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">g</span><span class="o">.</span><span class="n">dp</span> <span class="o">-</span> <span class="nb">sum</span><span class="p">(</span><span class="n">g</span><span class="o">.</span><span class="n">ad</span><span class="p">)</span>
</pre></div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.one_hot_alleles">
<code class="descname">one_hot_alleles</code><span class="sig-paren">(</span><em>num_alleles</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.one_hot_alleles"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.one_hot_alleles" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list containing the one-hot encoded representation of the called alleles.</p>
<p>This one-hot representation is the positional sum of the one-hot
encoding for each called allele.  For a biallelic variant, the
one-hot encoding for a reference allele is [1, 0] and the one-hot
encoding for an alternate allele is [0, 1].  Thus, with the
following variables:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">num_alleles</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">hom_ref</span> <span class="o">=</span> <span class="n">Genotype</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">het</span> <span class="o">=</span> <span class="n">Genotype</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">hom_var</span> <span class="o">=</span> <span class="n">Genotype</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>All the below statements are true:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">hom_ref</span><span class="o">.</span><span class="n">one_hot_alleles</span><span class="p">(</span><span class="n">num_alleles</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">het</span><span class="o">.</span><span class="n">one_hot_alleles</span><span class="p">(</span><span class="n">num_alleles</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
<span class="n">hom_var</span><span class="o">.</span><span class="n">one_hot_alleles</span><span class="p">(</span><span class="n">num_alleles</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
</pre></div>
</div>
<p>This function returns None if the genotype call is missing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>num_alleles</strong> (<em>int</em>) &#8211; number of possible alternate alleles</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.one_hot_genotype">
<code class="descname">one_hot_genotype</code><span class="sig-paren">(</span><em>num_genotypes</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.one_hot_genotype"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.one_hot_genotype" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a list containing the one-hot encoded representation of the genotype call.</p>
<p>A one-hot encoding is a vector with one &#8216;1&#8217; and many &#8216;0&#8217; values, like
[0, 0, 1, 0] or [1, 0, 0, 0].  This function is useful for transforming
the genotype call (gt) into a one-hot encoded array.  With the following
variables:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">num_genotypes</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">hom_ref</span> <span class="o">=</span> <span class="n">Genotype</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="n">het</span> <span class="o">=</span> <span class="n">Genotype</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="n">hom_var</span> <span class="o">=</span> <span class="n">Genotype</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
</pre></div>
</div>
<p>All the below statements are true:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">hom_ref</span><span class="o">.</span><span class="n">one_hot_genotype</span><span class="p">(</span><span class="n">num_genotypes</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">het</span><span class="o">.</span><span class="n">one_hot_genotype</span><span class="p">(</span><span class="n">num_genotypes</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]</span>
<span class="n">hom_var</span><span class="o">.</span><span class="n">one_hot_genotype</span><span class="p">(</span><span class="n">num_genotypes</span><span class="p">)</span> <span class="o">==</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>
</div>
<p>This function returns None if the genotype call is missing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>num_genotypes</strong> (<em>int</em>) &#8211; number of possible genotypes</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list of int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.representation.Genotype.p_ab">
<code class="descname">p_ab</code><span class="sig-paren">(</span><em>theta=0.5</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/hail/representation/genotype.html#Genotype.p_ab"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.representation.Genotype.p_ab" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the p-value associated with finding the given allele depth ratio.</p>
<p>This function uses a one-tailed binomial test.</p>
<p>This function returns None if the allelic depth (ad) is missing.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>theta</strong> (<em>float</em>) &#8211; null reference probability for binomial model</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">float or None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="attribute">
<dt id="hail.representation.Genotype.pl">
<code class="descname">pl</code><a class="headerlink" href="#hail.representation.Genotype.pl" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the phred-scaled genotype posterior likelihoods.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of int or None</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="hail.representation.Call.html" class="btn btn-neutral float-right" title="Call" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="hail.representation.AltAllele.html" class="btn btn-neutral" title="AltAllele" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Hail Team.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'devel',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      <script type="text/javascript" src="../../bootstrap.min.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
  
  
   

</body>
</html>