

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>HailContext &mdash; Hail</title>
  

  
  
    <link rel="shortcut icon" href="hail_logo_sq.png"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="navbar.css" type="text/css" />
  
    <link rel="stylesheet" href="_static/rtd_modifications.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Hail" href="index.html"/>
        <link rel="up" title="Python API" href="api.html"/>
        <link rel="next" title="VariantDataset" href="hail.VariantDataset.html"/>
        <link rel="prev" title="Python API" href="api.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>
  
  

</head>

<body class="wy-body-for-nav" role="document">

  <nav class="navbar navbar-default navbar-static-top" id="hail-navbar">
<div class="container-fluid" id="hail-container-fluid">
    <div class="navbar-header" id="hail-navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#hail-navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
   
      <a class="navbar-left" id="hail-navbar-brand" href="../index.html"><img alt="Hail" id="logo" src="hail-logo-cropped.png"></a>

    </div>

    <div class="collapse navbar-collapse" id="hail-navbar-collapse">
      <ul class="nav navbar-nav navbar-right">
          <li id="home" class="nav-item"><a href="../index.html">HOME</a></li>
          <li id="docs" class="nav-item"><a href="index.html">DOCS</a></li>
          <li id="forum" class="nav-item"><a href="http://discuss.hail.is">FORUM</a></li>
          <li id="chat" class="nav-item dropdown">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">CHAT<span class="caret"></span></a>
          <ul class="dropdown-menu dropdown-menu-left">
            <li><a class="dropdown-item" href="https://gitter.im/hail-is/hail">General</a></li>
            <li><a class="dropdown-item" href="https://gitter.im/hail-is/hail-dev">Developers</a></li>
          </ul>
        </li>
       <li id="code" class="nav-item"><a href="https://github.com/hail-is/hail">CODE</a></li>
      </ul>
    </div>
  </div>
</nav>

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Hail
          

          
          </a>

          
            
            
              <div class="version">
                devel
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="exprlang.html">Expression Language</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">Python API</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">HailContext</a></li>
<li class="toctree-l2"><a class="reference internal" href="hail.VariantDataset.html">VariantDataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="hail.KeyTable.html">KeyTable</a></li>
<li class="toctree-l2"><a class="reference internal" href="hail.TextTableConfig.html">TextTableConfig</a></li>
<li class="toctree-l2"><a class="reference internal" href="hail.KinshipMatrix.html">KinshipMatrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="representation/index.html">representation</a></li>
<li class="toctree-l2"><a class="reference internal" href="expr/index.html">expr</a></li>
<li class="toctree-l2"><a class="reference internal" href="utils/index.html">utils</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="annotationdb.html">Annotation Database</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_resources.html">Other Resources</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Hail</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="api.html">Python API</a> &raquo;</li>
        
      <li>HailContext</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/hail.HailContext.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

/* nice headers on first paragraph of info/warning boxes */
.admonition .first {
    margin: -12px;
    padding: 6px 12px;
    margin-bottom: 12px;
    color: #fff;
    line-height: 1;
    display: block;
}
.admonition.warning .first {
    background: #f0b37e;
}
.admonition.note .first {
    background: #6ab0de;
}
.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="hailcontext">
<h1>HailContext<a class="headerlink" href="#hailcontext" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="hail.HailContext">
<em class="property">class </em><code class="descclassname">hail.</code><code class="descname">HailContext</code><span class="sig-paren">(</span><em>sc=None</em>, <em>appName='Hail'</em>, <em>master=None</em>, <em>local='local[*]'</em>, <em>log='hail.log'</em>, <em>quiet=False</em>, <em>append=False</em>, <em>parquet_compression='snappy'</em>, <em>min_block_size=1</em>, <em>branching_factor=50</em>, <em>tmp_dir='/tmp'</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext" title="Permalink to this definition">¶</a></dt>
<dd><p>The main entry point for Hail functionality.</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">Only one Hail context may be running in a Python session at any time. If you
need to reconfigure settings, restart the Python session or use the <a class="reference internal" href="#hail.HailContext.stop" title="hail.HailContext.stop"><code class="xref py py-meth docutils literal"><span class="pre">HailContext.stop()</span></code></a> method.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>sc</strong> (<code class="xref py py-class docutils literal"><span class="pre">pyspark.SparkContext</span></code>) &#8211; spark context, will be auto-generated if None</li>
<li><strong>appName</strong> &#8211; Spark application identifier</li>
<li><strong>master</strong> &#8211; Spark cluster master</li>
<li><strong>local</strong> &#8211; local resources to use</li>
<li><strong>log</strong> &#8211; log path</li>
<li><strong>quiet</strong> &#8211; suppress log messages</li>
<li><strong>append</strong> &#8211; write to end of log file instead of overwriting</li>
<li><strong>parquet_compression</strong> &#8211; level of on-disk annotation compression</li>
<li><strong>min_block_size</strong> &#8211; minimum file split size in MB</li>
<li><strong>branching_factor</strong> &#8211; branching factor for tree aggregation</li>
<li><strong>tmp_dir</strong> &#8211; temporary directory for file merging</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><p class="first last"><strong>sc</strong> (<code class="xref py py-class docutils literal"><span class="pre">pyspark.SparkContext</span></code>) &#8211; Spark context</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Attributes</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.version" title="hail.HailContext.version"><code class="xref py py-obj docutils literal"><span class="pre">version</span></code></a></td>
<td>Return the version of Hail associated with this HailContext.</td>
</tr>
</tbody>
</table>
<p class="rubric">Methods</p>
<table border="1" class="longtable docutils">
<colgroup>
<col width="10%" />
<col width="90%" />
</colgroup>
<tbody valign="top">
<tr class="row-odd"><td><code class="xref py py-obj docutils literal"><span class="pre">__init__</span></code></td>
<td></td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.balding_nichols_model" title="hail.HailContext.balding_nichols_model"><code class="xref py py-obj docutils literal"><span class="pre">balding_nichols_model</span></code></a></td>
<td>Simulate a variant dataset using the Balding-Nichols model.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.dataframe_to_keytable" title="hail.HailContext.dataframe_to_keytable"><code class="xref py py-obj docutils literal"><span class="pre">dataframe_to_keytable</span></code></a></td>
<td>Convert Spark SQL DataFrame to key table.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.eval_expr" title="hail.HailContext.eval_expr"><code class="xref py py-obj docutils literal"><span class="pre">eval_expr</span></code></a></td>
<td>Evaluate an expression.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.eval_expr_typed" title="hail.HailContext.eval_expr_typed"><code class="xref py py-obj docutils literal"><span class="pre">eval_expr_typed</span></code></a></td>
<td>Evaluate an expression and return the result as well as its type.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.get_running" title="hail.HailContext.get_running"><code class="xref py py-obj docutils literal"><span class="pre">get_running</span></code></a></td>
<td>Return the running Hail context in this Python session.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.grep" title="hail.HailContext.grep"><code class="xref py py-obj docutils literal"><span class="pre">grep</span></code></a></td>
<td>Grep big files, like, really fast.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.import_bgen" title="hail.HailContext.import_bgen"><code class="xref py py-obj docutils literal"><span class="pre">import_bgen</span></code></a></td>
<td>Import .bgen file(s) as variant dataset.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.import_gen" title="hail.HailContext.import_gen"><code class="xref py py-obj docutils literal"><span class="pre">import_gen</span></code></a></td>
<td>Import .gen file(s) as variant dataset.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.import_keytable" title="hail.HailContext.import_keytable"><code class="xref py py-obj docutils literal"><span class="pre">import_keytable</span></code></a></td>
<td>Import delimited text file (text table) as key table.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.import_plink" title="hail.HailContext.import_plink"><code class="xref py py-obj docutils literal"><span class="pre">import_plink</span></code></a></td>
<td>Import PLINK binary file (BED, BIM, FAM) as variant dataset.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.import_vcf" title="hail.HailContext.import_vcf"><code class="xref py py-obj docutils literal"><span class="pre">import_vcf</span></code></a></td>
<td>Import VCF file(s) as variant dataset.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.index_bgen" title="hail.HailContext.index_bgen"><code class="xref py py-obj docutils literal"><span class="pre">index_bgen</span></code></a></td>
<td>Index .bgen files.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.read" title="hail.HailContext.read"><code class="xref py py-obj docutils literal"><span class="pre">read</span></code></a></td>
<td>Read .vds files as variant dataset.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.read_keytable" title="hail.HailContext.read_keytable"><code class="xref py py-obj docutils literal"><span class="pre">read_keytable</span></code></a></td>
<td>Read a KT file as key table.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.report" title="hail.HailContext.report"><code class="xref py py-obj docutils literal"><span class="pre">report</span></code></a></td>
<td>Print information and warnings about VCF + GEN import and deduplication.</td>
</tr>
<tr class="row-odd"><td><a class="reference internal" href="#hail.HailContext.stop" title="hail.HailContext.stop"><code class="xref py py-obj docutils literal"><span class="pre">stop</span></code></a></td>
<td>Shut down the Hail context.</td>
</tr>
<tr class="row-even"><td><a class="reference internal" href="#hail.HailContext.write_partitioning" title="hail.HailContext.write_partitioning"><code class="xref py py-obj docutils literal"><span class="pre">write_partitioning</span></code></a></td>
<td>Write partitioning.json.gz file for legacy VDS file.</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="hail.HailContext.balding_nichols_model">
<code class="descname">balding_nichols_model</code><span class="sig-paren">(</span><em>populations</em>, <em>samples</em>, <em>variants</em>, <em>npartitions=None</em>, <em>pop_dist=None</em>, <em>fst=None</em>, <em>af_dist=&lt;hail.stats.UniformDist instance&gt;</em>, <em>seed=0</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.balding_nichols_model"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.balding_nichols_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate a variant dataset using the Balding-Nichols model.</p>
<p><strong>Examples</strong></p>
<p>To generate a VDS with 3 populations, 100 samples in total, and 1000 variants:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vds</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">balding_nichols_model</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1000</span><span class="p">)</span>
</pre></div>
</div>
<p>To generate a VDS with 4 populations, 2000 samples, 5000 variants, 10 partitions, population distribution [0.1, 0.2, 0.3, 0.4], <span class="math">\(F_{ST}\)</span> values [.02, .06, .04, .12], ancestral allele frequencies drawn from a truncated beta distribution with a = .01 and b = .05 over the interval [0.05, 1], and random seed 1:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">from</span> <span class="nn">hail.stats</span> <span class="k">import</span> <span class="n">TruncatedBetaDist</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vds</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">balding_nichols_model</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">40</span><span class="p">,</span> <span class="mi">150</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span>
<span class="gp">... </span>                               <span class="n">pop_dist</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.4</span><span class="p">],</span>
<span class="gp">... </span>                               <span class="n">fst</span><span class="o">=</span><span class="p">[</span><span class="o">.</span><span class="mi">02</span><span class="p">,</span> <span class="o">.</span><span class="mi">06</span><span class="p">,</span> <span class="o">.</span><span class="mi">04</span><span class="p">,</span> <span class="o">.</span><span class="mi">12</span><span class="p">],</span>
<span class="gp">... </span>                               <span class="n">af_dist</span><span class="o">=</span><span class="n">TruncatedBetaDist</span><span class="p">(</span><span class="n">a</span><span class="o">=</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">b</span><span class="o">=</span><span class="mf">2.0</span><span class="p">,</span> <span class="n">minVal</span><span class="o">=</span><span class="mf">0.05</span><span class="p">,</span> <span class="n">maxVal</span><span class="o">=</span><span class="mf">1.0</span><span class="p">),</span>
<span class="gp">... </span>                               <span class="n">seed</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Notes</strong></p>
<p>Hail is able to randomly generate a VDS using the Balding-Nichols model.</p>
<ul class="simple">
<li><span class="math">\(K\)</span> populations are labeled by integers 0, 1, ..., K - 1</li>
<li><span class="math">\(N\)</span> samples are named by strings 0, 1, ..., N - 1</li>
<li><span class="math">\(M\)</span> variants are defined as <code class="docutils literal"><span class="pre">1:1:A:C</span></code>, <code class="docutils literal"><span class="pre">1:2:A:C</span></code>, ..., <code class="docutils literal"><span class="pre">1:M:A:C</span></code></li>
<li>The default ancestral frequency distribution <span class="math">\(P_0\)</span> is uniform on [0.1, 0.9]. Options are UniformDist(minVal, maxVal), BetaDist(a, b), and TruncatedBetaDist(a, b, minVal, maxVal). All three classes are located in hail.stats.</li>
<li>The population distribution <span class="math">\(\pi\)</span> defaults to uniform</li>
<li>The <span class="math">\(F_{ST}\)</span> values default to 0.1</li>
<li>The number of partitions defaults to one partition per million genotypes (i.e., samples * variants / 10^6) or 8, whichever is larger</li>
</ul>
<p>The Balding-Nichols model models genotypes of individuals from a structured population comprising <span class="math">\(K\)</span> homogeneous subpopulations
that have each diverged from a single ancestral population (a <cite>star phylogeny</cite>). We take <span class="math">\(N\)</span> samples and <span class="math">\(M\)</span> bi-allelic variants in perfect
linkage equilibrium. The relative sizes of the subpopulations are given by a probability vector <span class="math">\(\pi\)</span>; the ancestral allele frequencies are
drawn independently from a frequency spectrum <span class="math">\(P_0\)</span>; the subpopulations have diverged with possibly different <span class="math">\(F_{ST}\)</span> parameters <span class="math">\(F_k\)</span>
(here and below, lowercase indices run over a range bounded by the corresponding uppercase parameter, e.g. <span class="math">\(k = 1, \ldots, K\)</span>).
For each variant, the subpopulation allele frequencies are drawn a <a class="reference external" href="https://en.wikipedia.org/wiki/Beta_distribution">beta distribution</a>, a useful continuous approximation of
the effect of genetic drift. We denote the individual subpopulation memberships by <span class="math">\(k_n\)</span>, the ancestral allele frequences by <span class="math">\(p_{0, m}\)</span>,
the subpopulation allele frequencies by <span class="math">\(p_{k, m}\)</span>, and the genotypes by <span class="math">\(g_{n, m}\)</span>. The generative model in then given by:</p>
<div class="math">
\[ \begin{align}\begin{aligned}k_n \,&amp;\sim\, \pi\\p_{0,m}\,&amp;\sim\, P_0\\p_{k,m}\mid p_{0,m}\,&amp;\sim\, \mathrm{Beta}(\mu = p_{0,m},\, \sigma^2 = F_k p_{0,m}(1 - p_{0,m}))\\g_{n,m}\mid k_n, p_{k, m} \,&amp;\sim\, \mathrm{Binomial}(2, p_{k_n, m})\end{aligned}\end{align} \]</div>
<p>We have parametrized the beta distribution by its mean and variance; the usual parameters are <span class="math">\(a = (1 - p)(1 - F)/F,\; b = p(1-F)/F\)</span> with <span class="math">\(F = F_k,\; p = p_{0,m}\)</span>.</p>
<p><strong>Annotations</strong></p>
<p><a class="reference internal" href="#hail.HailContext.balding_nichols_model" title="hail.HailContext.balding_nichols_model"><code class="xref py py-meth docutils literal"><span class="pre">balding_nichols_model()</span></code></a> adds the following global, sample, and variant annotations:</p>
<blockquote>
<div><ul class="simple">
<li><strong>global.nPops</strong> (<em>Int</em>) &#8211; Number of populations</li>
<li><strong>global.nSamples</strong> (<em>Int</em>) &#8211; Number of samples</li>
<li><strong>global.nVariants</strong> (<em>Int</em>) &#8211; Number of variants</li>
<li><strong>global.popDist</strong> (<em>Array[Double]</em>) &#8211; Normalized population distribution indexed by population</li>
<li><strong>global.Fst</strong> (<em>Array[Double]</em>) &#8211; <span class="math">\(F_{ST}\)</span> values indexed by population</li>
<li><strong>global.seed</strong> (<em>Int</em>) &#8211; Random seed</li>
<li><strong>global.ancestralAFDist</strong> (<em>Struct</em>) &#8211; Description of the ancestral allele frequency distribution</li>
<li><strong>sa.pop</strong> (<em>Int</em>) &#8211; Population of sample</li>
<li><strong>va.ancestralAF</strong> (<em>Double</em>) &#8211; Ancestral allele frequency</li>
<li><strong>va.AF</strong> (<em>Array[Double]</em>) &#8211; Allele frequency indexed by population</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>populations</strong> (<em>int</em>) &#8211; Number of populations.</li>
<li><strong>samples</strong> (<em>int</em>) &#8211; Number of samples.</li>
<li><strong>variants</strong> (<em>int</em>) &#8211; Number of variants.</li>
<li><strong>npartitions</strong> (<em>int</em>) &#8211; Number of partitions.</li>
<li><strong>pop_dist</strong> (<em>array of float</em><em> or </em><em>None</em>) &#8211; Unnormalized population distribution</li>
<li><strong>fst</strong> (<em>array of float</em><em> or </em><em>None</em>) &#8211; <span class="math">\(F_{ST}\)</span> values</li>
<li><strong>af_dist</strong> (<code class="xref py py-class docutils literal"><span class="pre">UniformDist</span></code> or <code class="xref py py-class docutils literal"><span class="pre">BetaDist</span></code> or <code class="xref py py-class docutils literal"><span class="pre">TruncatedBetaDist</span></code>) &#8211; Ancestral allele frequency distribution</li>
<li><strong>seed</strong> (<em>int</em>) &#8211; Random seed.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Variant dataset simulated using the Balding-Nichols model.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.dataframe_to_keytable">
<code class="descname">dataframe_to_keytable</code><span class="sig-paren">(</span><em>df</em>, <em>keys=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.dataframe_to_keytable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.dataframe_to_keytable" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert Spark SQL DataFrame to key table.</p>
<p>Spark SQL data types are converted to Hail types in the obvious way as follows:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>BooleanType =&gt; Boolean
IntegerType =&gt; Int
LongType =&gt; Long
FloatType =&gt; Float
DoubleType =&gt; Double
StringType =&gt; String
BinaryType =&gt; Binary
ArrayType =&gt; Array
StructType =&gt; Struct
</pre></div>
</div>
<p>Unlisted Spark SQL data types are currently unsupported.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>keys</strong> (<em>list of string</em>) &#8211; List of key column names.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Key table constructed from the Spark SQL DataFrame.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="hail.KeyTable.html#hail.KeyTable" title="hail.KeyTable"><code class="xref py py-class docutils literal"><span class="pre">KeyTable</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.eval_expr">
<code class="descname">eval_expr</code><span class="sig-paren">(</span><em>expr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.eval_expr"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.eval_expr" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate an expression.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expr</strong> (<em>str</em>) &#8211; Expression to evaluate.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">annotation</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.eval_expr_typed">
<code class="descname">eval_expr_typed</code><span class="sig-paren">(</span><em>expr</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.eval_expr_typed"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.eval_expr_typed" title="Permalink to this definition">¶</a></dt>
<dd><p>Evaluate an expression and return the result as well as its type.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>expr</strong> (<em>str</em>) &#8211; Expression to evaluate.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">(annotation, <a class="reference internal" href="expr/hail.expr.Type.html#hail.expr.Type" title="hail.expr.Type"><code class="xref py py-class docutils literal"><span class="pre">Type</span></code></a>)</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="staticmethod">
<dt id="hail.HailContext.get_running">
<em class="property">static </em><code class="descname">get_running</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.get_running"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.get_running" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the running Hail context in this Python session.</p>
<p><strong>Example</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">HailContext</span><span class="p">()</span>  <span class="c1"># oops! Forgot to bind to &#39;hc&#39;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span> <span class="o">=</span> <span class="n">HailContext</span><span class="o">.</span><span class="n">get_running</span><span class="p">()</span>  <span class="c1"># recovery</span>
</pre></div>
</div>
<p>Useful to recover a Hail context that has been created but is unbound.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body">Current Hail context.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="#hail.HailContext" title="hail.HailContext"><code class="xref py py-class docutils literal"><span class="pre">HailContext</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.grep">
<code class="descname">grep</code><span class="sig-paren">(</span><em>regex</em>, <em>path</em>, <em>max_count=100</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.grep"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.grep" title="Permalink to this definition">¶</a></dt>
<dd><p>Grep big files, like, really fast.</p>
<p><strong>Examples</strong></p>
<p>Print all lines containing the string <code class="docutils literal"><span class="pre">hello</span></code> in <em>file.txt</em>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">grep</span><span class="p">(</span><span class="s1">&#39;hello&#39;</span><span class="p">,</span><span class="s1">&#39;data/file.txt&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Print all lines containing digits in <em>file1.txt</em> and <em>file2.txt</em>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">grep</span><span class="p">(</span><span class="s1">&#39;\d&#39;</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;data/file1.txt&#39;</span><span class="p">,</span><span class="s1">&#39;data/file2.txt&#39;</span><span class="p">])</span>
</pre></div>
</div>
<p><strong>Background</strong></p>
<p><a class="reference internal" href="#hail.HailContext.grep" title="hail.HailContext.grep"><code class="xref py py-meth docutils literal"><span class="pre">grep()</span></code></a> mimics the basic functionality of Unix <code class="docutils literal"><span class="pre">grep</span></code> in parallel, printing results to screen. This command is provided as a convenience to those in the statistical genetics community who often search enormous text files like VCFs. Find background on regular expressions at <a class="reference external" href="http://regexr.com/">RegExr</a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>regex</strong> (<em>str</em>) &#8211; The regular expression to match.</li>
<li><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; The files to search.</li>
<li><strong>max_count</strong> (<em>int</em>) &#8211; The maximum number of matches to return.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.import_bgen">
<code class="descname">import_bgen</code><span class="sig-paren">(</span><em>path</em>, <em>tolerance=0.2</em>, <em>sample_file=None</em>, <em>npartitions=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.import_bgen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.import_bgen" title="Permalink to this definition">¶</a></dt>
<dd><p>Import .bgen file(s) as variant dataset.</p>
<p><strong>Examples</strong></p>
<p>Importing a BGEN file as a VDS (assuming it has already been indexed).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vds</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">import_bgen</span><span class="p">(</span><span class="s2">&quot;data/example3.bgen&quot;</span><span class="p">,</span> <span class="n">sample_file</span><span class="o">=</span><span class="s2">&quot;data/example3.sample&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Notes</strong></p>
<p>Hail supports importing data in the BGEN file format. For more information on the BGEN file format,
see <a class="reference external" href="http://www.well.ox.ac.uk/~gav/bgen_format/bgen_format_v1.1.html">here</a>. Note that only v1.1 BGEN files
are supported at this time.</p>
<p>Before importing, ensure that:
- Files reside in Hadoop file system.
- The sample file has the same number of samples as the BGEN file.
- No duplicate sample IDs are present.</p>
<p>To load multiple files at the same time, use <a class="reference internal" href="hadoop_glob_patterns.html#sec-hadoop-glob"><span class="std std-ref">Hadoop Glob Patterns</span></a>.</p>
<p id="dosagefilters"><strong>Dosage representation</strong>:</p>
<p>Since dosages are understood as genotype probabilities, <a class="reference internal" href="#hail.HailContext.import_bgen" title="hail.HailContext.import_bgen"><code class="xref py py-meth docutils literal"><span class="pre">import_bgen()</span></code></a> automatically sets to missing those genotypes for which the sum of the dosages is a distance greater than the <code class="docutils literal"><span class="pre">tolerance</span></code> parameter from 1.0.  The default tolerance is 0.2, so a genotype with sum .79 or 1.21 is filtered out, whereas a genotype with sum .8 or 1.2 remains.</p>
<p><a class="reference internal" href="#hail.HailContext.import_gen" title="hail.HailContext.import_gen"><code class="xref py py-meth docutils literal"><span class="pre">import_gen()</span></code></a> normalizes all dosages to sum to 1.0. Therefore, an input dosage of (0.98, 0.0, 0.0) will be stored as (1.0, 0.0, 0.0) in Hail.</p>
<p>Even when the dosages sum to 1.0, Hail may store slightly different values than the original GEN file (maximum observed difference is 3E-4).</p>
<p><strong>Annotations</strong></p>
<p><a class="reference internal" href="#hail.HailContext.import_bgen" title="hail.HailContext.import_bgen"><code class="xref py py-meth docutils literal"><span class="pre">import_bgen()</span></code></a> adds the following variant annotations:</p>
<blockquote>
<div><ul class="simple">
<li><strong>va.varid</strong> (<em>String</em>) &#8211; 2nd column of .gen file if chromosome present, otherwise 1st column.</li>
<li><strong>va.rsid</strong> (<em>String</em>) &#8211; 3rd column of .gen file if chromosome present, otherwise 2nd column.</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; .bgen files to import.</li>
<li><strong>tolerance</strong> (<em>float</em>) &#8211; If the sum of the dosages for a
genotype differ from 1.0 by more than the tolerance, set
the genotype to missing.</li>
<li><strong>sample_file</strong> (<em>str</em><em> or </em><em>None</em>) &#8211; Sample file.</li>
<li><strong>npartitions</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; Number of partitions.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Variant dataset imported from .bgen file.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.import_gen">
<code class="descname">import_gen</code><span class="sig-paren">(</span><em>path</em>, <em>sample_file=None</em>, <em>tolerance=0.2</em>, <em>npartitions=None</em>, <em>chromosome=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.import_gen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.import_gen" title="Permalink to this definition">¶</a></dt>
<dd><p>Import .gen file(s) as variant dataset.</p>
<p><strong>Examples</strong></p>
<p>Read a .gen file and a .sample file and write to a .vds file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">hc</span><span class="o">.</span><span class="n">import_gen</span><span class="p">(</span><span class="s1">&#39;data/example.gen&#39;</span><span class="p">,</span> <span class="n">sample_file</span><span class="o">=</span><span class="s1">&#39;data/example.sample&#39;</span><span class="p">)</span>
<span class="gp">... </span>   <span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;output/gen_example1.vds&#39;</span><span class="p">))</span>
</pre></div>
</div>
<p>Load multiple files at the same time with <a class="reference internal" href="hadoop_glob_patterns.html#sec-hadoop-glob"><span class="std std-ref">Hadoop glob patterns</span></a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">hc</span><span class="o">.</span><span class="n">import_gen</span><span class="p">(</span><span class="s1">&#39;data/example.chr*.gen&#39;</span><span class="p">,</span> <span class="n">sample_file</span><span class="o">=</span><span class="s1">&#39;data/example.sample&#39;</span><span class="p">)</span>
<span class="gp">... </span>   <span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;output/gen_example2.vds&#39;</span><span class="p">))</span>
</pre></div>
</div>
<p><strong>Notes</strong></p>
<p>For more information on the .gen file format, see <a class="reference external" href="http://www.stats.ox.ac.uk/%7Emarchini/software/gwas/file_format.html#mozTocId40300">here</a>.</p>
<p>To ensure that the .gen file(s) and .sample file are correctly prepared for import:</p>
<ul class="simple">
<li>If there are only 5 columns before the start of the dosage data (chromosome field is missing), you must specify the chromosome using the <code class="docutils literal"><span class="pre">chromosome</span></code> parameter</li>
<li>No duplicate sample IDs are allowed</li>
</ul>
<p>The first column in the .sample file is used as the sample ID <code class="docutils literal"><span class="pre">s</span></code>.</p>
<p>Also, see section in <a class="reference internal" href="#hail.HailContext.import_bgen" title="hail.HailContext.import_bgen"><code class="xref py py-meth docutils literal"><span class="pre">import_bgen()</span></code></a> linked <a class="reference internal" href="#dosagefilters"><span class="std std-ref">here</span></a> for information about Hail&#8217;s dosage representation.</p>
<p><strong>Annotations</strong></p>
<p><a class="reference internal" href="#hail.HailContext.import_gen" title="hail.HailContext.import_gen"><code class="xref py py-meth docutils literal"><span class="pre">import_gen()</span></code></a> adds the following variant annotations:</p>
<blockquote>
<div><ul class="simple">
<li><strong>va.varid</strong> (<em>String</em>) &#8211; 2nd column of .gen file if chromosome present, otherwise 1st column.</li>
<li><strong>va.rsid</strong> (<em>String</em>) &#8211; 3rd column of .gen file if chromosome present, otherwise 2nd column.</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; .gen files to import.</li>
<li><strong>sample_file</strong> (<em>str</em>) &#8211; The sample file.</li>
<li><strong>tolerance</strong> (<em>float</em>) &#8211; If the sum of the dosages for a genotype differ from 1.0 by more than the tolerance, set the genotype to missing.</li>
<li><strong>npartitions</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; Number of partitions.</li>
<li><strong>chromosome</strong> (<em>str</em><em> or </em><em>None</em>) &#8211; Chromosome if not listed in the .gen file.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Variant dataset imported from .gen and .sample files.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.import_keytable">
<code class="descname">import_keytable</code><span class="sig-paren">(</span><em>path</em>, <em>key=[]</em>, <em>npartitions=None</em>, <em>config=&lt;hail.utils.TextTableConfig object&gt;</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.import_keytable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.import_keytable" title="Permalink to this definition">¶</a></dt>
<dd><p>Import delimited text file (text table) as key table.</p>
<p>The resulting key table will have no key columns, use <a class="reference internal" href="hail.KeyTable.html#hail.KeyTable.key_by" title="hail.KeyTable.key_by"><code class="xref py py-meth docutils literal"><span class="pre">KeyTable.key_by()</span></code></a>
to specify keys.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; files to import.</li>
<li><strong>key</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; List of key columns.</li>
<li><strong>npartitions</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; Number of partitions.</li>
<li><strong>config</strong> (<a class="reference internal" href="hail.TextTableConfig.html#hail.TextTableConfig" title="hail.TextTableConfig"><code class="xref py py-class docutils literal"><span class="pre">TextTableConfig</span></code></a>) &#8211; Configuration options for importing text files</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Key table constructed from text table.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.KeyTable.html#hail.KeyTable" title="hail.KeyTable"><code class="xref py py-class docutils literal"><span class="pre">KeyTable</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.import_plink">
<code class="descname">import_plink</code><span class="sig-paren">(</span><em>bed</em>, <em>bim</em>, <em>fam</em>, <em>npartitions=None</em>, <em>delimiter='\\\\s+'</em>, <em>missing='NA'</em>, <em>quantpheno=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.import_plink"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.import_plink" title="Permalink to this definition">¶</a></dt>
<dd><p>Import PLINK binary file (BED, BIM, FAM) as variant dataset.</p>
<p><strong>Examples</strong></p>
<p>Import data from a PLINK binary file:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vds</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">import_plink</span><span class="p">(</span><span class="n">bed</span><span class="o">=</span><span class="s2">&quot;data/test.bed&quot;</span><span class="p">,</span>
<span class="gp">... </span>                      <span class="n">bim</span><span class="o">=</span><span class="s2">&quot;data/test.bim&quot;</span><span class="p">,</span>
<span class="gp">... </span>                      <span class="n">fam</span><span class="o">=</span><span class="s2">&quot;data/test.fam&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Notes</strong></p>
<p>Only binary SNP-major mode files can be read into Hail. To convert your file from individual-major mode to SNP-major mode, use PLINK to read in your fileset and use the <code class="docutils literal"><span class="pre">--make-bed</span></code> option.</p>
<p>The centiMorgan position is not currently used in Hail (Column 3 in BIM file).</p>
<p>The ID (<code class="docutils literal"><span class="pre">s</span></code>) used by Hail is the individual ID (column 2 in FAM file).</p>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">No duplicate individual IDs are allowed.</p>
</div>
<p>Chromosome names (Column 1) are automatically converted in the following cases:</p>
<blockquote>
<div><ul class="simple">
<li>23 =&gt; &#8220;X&#8221;</li>
<li>24 =&gt; &#8220;Y&#8221;</li>
<li>25 =&gt; &#8220;X&#8221;</li>
<li>26 =&gt; &#8220;MT&#8221;</li>
</ul>
</div></blockquote>
<p><strong>Annotations</strong></p>
<p><a class="reference internal" href="#hail.HailContext.import_plink" title="hail.HailContext.import_plink"><code class="xref py py-meth docutils literal"><span class="pre">import_plink()</span></code></a> adds the following annotations:</p>
<blockquote>
<div><ul class="simple">
<li><strong>va.rsid</strong> (<em>String</em>) &#8211; Column 2 in the BIM file.</li>
<li><strong>sa.famID</strong> (<em>String</em>) &#8211; Column 1 in the FAM file. Set to missing if ID equals &#8220;0&#8221;.</li>
<li><strong>sa.patID</strong> (<em>String</em>) &#8211; Column 3 in the FAM file. Set to missing if ID equals &#8220;0&#8221;.</li>
<li><strong>sa.matID</strong> (<em>String</em>) &#8211; Column 4 in the FAM file. Set to missing if ID equals &#8220;0&#8221;.</li>
<li><strong>sa.isFemale</strong> (<em>String</em>) &#8211; Column 5 in the FAM file. Set to missing if value equals &#8220;-9&#8221;, &#8220;0&#8221;, or &#8220;N/A&#8221;.
Set to true if value equals &#8220;2&#8221;. Set to false if value equals &#8220;1&#8221;.</li>
<li><strong>sa.isCase</strong> (<em>String</em>) &#8211; Column 6 in the FAM file. Only present if <code class="docutils literal"><span class="pre">quantpheno</span></code> equals False.
Set to missing if value equals &#8220;-9&#8221;, &#8220;0&#8221;, &#8220;N/A&#8221;, or the value specified by <code class="docutils literal"><span class="pre">missing</span></code>.
Set to true if value equals &#8220;2&#8221;. Set to false if value equals &#8220;1&#8221;.</li>
<li><strong>sa.qPheno</strong> (<em>String</em>) &#8211; Column 6 in the FAM file. Only present if <code class="docutils literal"><span class="pre">quantpheno</span></code> equals True.
Set to missing if value equals <code class="docutils literal"><span class="pre">missing</span></code>.</li>
</ul>
</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>bed</strong> (<em>str</em>) &#8211; PLINK BED file.</li>
<li><strong>bim</strong> (<em>str</em>) &#8211; PLINK BIM file.</li>
<li><strong>fam</strong> (<em>str</em>) &#8211; PLINK FAM file.</li>
<li><strong>npartitions</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; Number of partitions.</li>
<li><strong>missing</strong> (<em>str</em>) &#8211; The string used to denote missing values <strong>only</strong> for the phenotype field. This is in addition to &#8220;-9&#8221;, &#8220;0&#8221;, and &#8220;N/A&#8221; for case-control phenotypes.</li>
<li><strong>delimiter</strong> (<em>str</em>) &#8211; FAM file field delimiter regex.</li>
<li><strong>quantpheno</strong> (<em>bool</em>) &#8211; If True, FAM phenotype is interpreted as quantitative.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Variant dataset imported from PLINK binary file.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.import_vcf">
<code class="descname">import_vcf</code><span class="sig-paren">(</span><em>path</em>, <em>force=False</em>, <em>force_bgz=False</em>, <em>header_file=None</em>, <em>npartitions=None</em>, <em>sites_only=False</em>, <em>store_gq=False</em>, <em>pp_as_pl=False</em>, <em>skip_bad_ad=False</em>, <em>generic=False</em>, <em>call_fields=[]</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.import_vcf"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.import_vcf" title="Permalink to this definition">¶</a></dt>
<dd><p>Import VCF file(s) as variant dataset.</p>
<p><strong>Examples</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vds</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">import_vcf</span><span class="p">(</span><span class="s1">&#39;data/example2.vcf.bgz&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>Notes</strong></p>
<p>Hail is designed to be maximally compatible with files in the <a class="reference external" href="https://samtools.github.io/hts-specs/VCFv4.2.pdf">VCF v4.2 spec</a>.</p>
<p><a class="reference internal" href="#hail.HailContext.import_vcf" title="hail.HailContext.import_vcf"><code class="xref py py-meth docutils literal"><span class="pre">import_vcf()</span></code></a> takes a list of VCF files to load. All files must have the same header and the same set of samples in the same order
(e.g., a variant dataset split by chromosome). Files can be specified as <a class="reference internal" href="hadoop_glob_patterns.html#sec-hadoop-glob"><span class="std std-ref">Hadoop glob patterns</span></a>.</p>
<p>Ensure that the VCF file is correctly prepared for import: VCFs should either be uncompressed (<em>.vcf</em>) or block compressed
(<em>.vcf.bgz</em>).  If you have a large compressed VCF that ends in <em>.vcf.gz</em>, it is likely that the file is actually block-compressed,
and you should rename the file to &#8221;.vcf.bgz&#8221; accordingly. If you actually have a standard gzipped file, it is possible to import
it to Hail using the <code class="docutils literal"><span class="pre">force</span></code> optional parameter. However, this is not recommended &#8211; all parsing will have to take place on one node because
gzip decompression is not parallelizable. In this case, import could take significantly longer.</p>
<p>If <code class="docutils literal"><span class="pre">generic</span></code> equals False (default), Hail makes certain assumptions about the genotype fields, see <a class="reference internal" href="representation/hail.representation.Genotype.html#hail.representation.Genotype" title="hail.representation.Genotype"><code class="xref py py-class docutils literal"><span class="pre">Representation</span></code></a>. On import, Hail filters
(sets to no-call) any genotype that violates these assumptions. Hail interprets the format fields: GT, AD, OD, DP, GQ, PL; all others are
silently dropped.</p>
<p>If <code class="docutils literal"><span class="pre">generic</span></code> equals True, the genotype schema is a <code class="xref py py-class docutils literal"><span class="pre">TStruct</span></code> with field names equal to the IDs of the FORMAT fields.
The <code class="docutils literal"><span class="pre">GT</span></code> field is automatically read in as a <code class="xref py py-class docutils literal"><span class="pre">TCall</span></code> type. To specify additional fields to import as a
<code class="xref py py-class docutils literal"><span class="pre">TCall</span></code> type, use the <code class="docutils literal"><span class="pre">call_fields</span></code> parameter. All other fields are imported as the type specified in the FORMAT header field.</p>
<p>An example genotype schema after importing a VCF with <code class="docutils literal"><span class="pre">generic=True</span></code> is</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>Struct {
    GT: Call,
    AD: Array[Int],
    DP: Int,
    GQ: Int,
    PL: Array[Int]
}
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<ul class="last simple">
<li>The variant dataset generated with <code class="docutils literal"><span class="pre">generic=True</span></code> will have significantly slower performance.</li>
<li>Not all <a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a> methods will work with a generic genotype schema.</li>
</ul>
</div>
<p><a class="reference internal" href="#hail.HailContext.import_vcf" title="hail.HailContext.import_vcf"><code class="xref py py-meth docutils literal"><span class="pre">import_vcf()</span></code></a> does not perform deduplication - if the provided VCF(s) contain multiple records with the same chrom, pos, ref, alt, all
these records will be imported and will not be collapsed into a single variant.</p>
<p>Since Hail&#8217;s genotype representation does not yet support ploidy other than 2,
this method imports haploid genotypes as diploid. If <code class="docutils literal"><span class="pre">generic=False</span></code>, Hail fills in missing indices
in PL / PP arrays with 1000 to support the standard VCF / VDS &#8220;genotype schema.</p>
<p>Below are two example haploid genotypes and diploid equivalents that Hail sees.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>Haploid:     1:0,6:7:70:70,0
Imported as: 1/1:0,6:7:70:70,1000,0

Haploid:     2:0,0,9:9:24:24,40,0
Imported as: 2/2:0,0,9:9:24:24,1000,40,1000:1000:0
</pre></div>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p>Using the <strong>FILTER</strong> field:</p>
<p>The information in the FILTER field of a VCF is contained in the <code class="docutils literal"><span class="pre">va.filters</span></code> annotation.
This annotation is a <code class="docutils literal"><span class="pre">Set</span></code> and can be queried for filter membership with expressions 
like <code class="docutils literal"><span class="pre">va.filters.contains(&quot;VQSRTranche99.5...&quot;)</span></code>. Variants that are flagged as &#8220;PASS&#8221; 
will have no filters applied; for these variants, <code class="docutils literal"><span class="pre">va.filters.isEmpty()</span></code> is true. Thus, 
filtering to PASS variants can be done with <a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset.filter_variants_expr" title="hail.VariantDataset.filter_variants_expr"><code class="xref py py-meth docutils literal"><span class="pre">VariantDataset.filter_variants_expr()</span></code></a>
as follows:</p>
<div class="last highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">pass_vds</span> <span class="o">=</span> <span class="n">vds</span><span class="o">.</span><span class="n">filter_variants_expr</span><span class="p">(</span><span class="s1">&#39;va.filters.isEmpty()&#39;</span><span class="p">,</span> <span class="n">keep</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p><strong>Annotations</strong></p>
<ul class="simple">
<li><strong>va.filters</strong> (<em>Set[String]</em>) &#8211; Set containing all filters applied to a variant.</li>
<li><strong>va.rsid</strong> (<em>String</em>) &#8211; rsID of the variant.</li>
<li><strong>va.qual</strong> (<em>Double</em>) &#8211; Floating-point number in the QUAL field.</li>
<li><strong>va.info</strong> (<em>Struct</em>) &#8211; All INFO fields defined in the VCF header can be found in the
struct <code class="docutils literal"><span class="pre">va.info</span></code>. Data types match the type specified in the VCF header, and if</li>
</ul>
<blockquote>
<div>the declared <code class="docutils literal"><span class="pre">Number</span></code> is not 1, the result will be stored as an array.</div></blockquote>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; VCF file(s) to read.</li>
<li><strong>force</strong> (<em>bool</em>) &#8211; If True, load .gz files serially. This means that no downstream operations
can be parallelized, so using this mode is strongly discouraged for VCFs larger than a few MB.</li>
<li><strong>force_bgz</strong> (<em>bool</em>) &#8211; If True, load .gz files as blocked gzip files (BGZF)</li>
<li><strong>header_file</strong> (<em>str</em><em> or </em><em>None</em>) &#8211; File to load VCF header from.  If not specified, the first file in path is used.</li>
<li><strong>npartitions</strong> (<em>int</em><em> or </em><em>None</em>) &#8211; Number of partitions.</li>
<li><strong>sites_only</strong> (<em>bool</em>) &#8211; If True, create sites-only
variant dataset.  Don&#8217;t load sample ids, sample annotations
or genotypes.</li>
<li><strong>store_gq</strong> (<em>bool</em>) &#8211; If True, store GQ FORMAT field instead of computing from PL. Only applies if <code class="docutils literal"><span class="pre">generic=False</span></code>.</li>
<li><strong>pp_as_pl</strong> (<em>bool</em>) &#8211; If True, store PP FORMAT field as PL.  EXPERIMENTAL. Only applies if <code class="docutils literal"><span class="pre">generic=False</span></code>.</li>
<li><strong>skip_bad_ad</strong> (<em>bool</em>) &#8211; If True, set AD FORMAT field with
wrong number of elements to missing, rather than setting
the entire genotype to missing. Only applies if <code class="docutils literal"><span class="pre">generic=False</span></code>.</li>
<li><strong>generic</strong> (<em>bool</em>) &#8211; If True, read the genotype with a generic schema.</li>
<li><strong>call_fields</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; FORMAT fields in VCF to treat as a <code class="xref py py-class docutils literal"><span class="pre">TCall</span></code>. Only applies if <code class="docutils literal"><span class="pre">generic=True</span></code>.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Variant dataset imported from VCF file(s)</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.index_bgen">
<code class="descname">index_bgen</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.index_bgen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.index_bgen" title="Permalink to this definition">¶</a></dt>
<dd><p>Index .bgen files. <a class="reference internal" href="#hail.HailContext.import_bgen" title="hail.HailContext.import_bgen"><code class="xref py py-meth docutils literal"><span class="pre">HailContext.import_bgen()</span></code></a> cannot run without these indices.</p>
<p><strong>Example</strong></p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vds</span> <span class="o">=</span> <span class="n">hc</span><span class="o">.</span><span class="n">index_bgen</span><span class="p">(</span><span class="s2">&quot;data/example3.bgen&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">While this method parallelizes over a list of BGEN files, each file is
indexed serially by one core. Indexing several BGEN files on a large
cluster is a waste of resources, so indexing should generally be done
as a one-time step separately from large analyses.</p>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; .bgen files to index.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.read">
<code class="descname">read</code><span class="sig-paren">(</span><em>path</em>, <em>sites_only=False</em>, <em>samples_only=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.read"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.read" title="Permalink to this definition">¶</a></dt>
<dd><p>Read .vds files as variant dataset.</p>
<p>When loading multiple .vds files, they must have the same
sample IDs, genotype schema, split status and variant metadata.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>path</strong> (<em>str</em><em> or </em><em>list of str</em>) &#8211; .vds files to read.</li>
<li><strong>sites_only</strong> (<em>bool</em>) &#8211; If True, create sites-only variant
dataset.  Don&#8217;t load sample ids, sample annotations
or gneotypes.</li>
<li><strong>samples_only</strong> (<em>bool</em>) &#8211; If True, create samples-only variant
dataset (no variants or genotypes).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">Variant dataset read from disk.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last"><a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset" title="hail.VariantDataset"><code class="xref py py-class docutils literal"><span class="pre">VariantDataset</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.read_keytable">
<code class="descname">read_keytable</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.read_keytable"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.read_keytable" title="Permalink to this definition">¶</a></dt>
<dd><p>Read a KT file as key table.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<em>str</em>) &#8211; KT file to read.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">Key table read from disk.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><a class="reference internal" href="hail.KeyTable.html#hail.KeyTable" title="hail.KeyTable"><code class="xref py py-class docutils literal"><span class="pre">KeyTable</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.report">
<code class="descname">report</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.report" title="Permalink to this definition">¶</a></dt>
<dd><p>Print information and warnings about VCF + GEN import and deduplication.</p>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.stop">
<code class="descname">stop</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.stop"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.stop" title="Permalink to this definition">¶</a></dt>
<dd><p>Shut down the Hail context.</p>
<p>It is not possible to have multiple Hail contexts running in a
single Python session, so use this if you need to reconfigure the Hail
context. Note that this also stops a running Spark context.</p>
</dd></dl>

<dl class="attribute">
<dt id="hail.HailContext.version">
<code class="descname">version</code><a class="headerlink" href="#hail.HailContext.version" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the version of Hail associated with this HailContext.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">str</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="hail.HailContext.write_partitioning">
<code class="descname">write_partitioning</code><span class="sig-paren">(</span><em>path</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/hail/context.html#HailContext.write_partitioning"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#hail.HailContext.write_partitioning" title="Permalink to this definition">¶</a></dt>
<dd><p>Write partitioning.json.gz file for legacy VDS file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>path</strong> (<em>str</em>) &#8211; path to VDS file.</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="hail.VariantDataset.html" class="btn btn-neutral float-right" title="VariantDataset" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="api.html" class="btn btn-neutral" title="Python API" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Hail Team.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'devel',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      <script type="text/javascript" src="../bootstrap.min.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
  
  
   

</body>
</html>