

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Querying using SQL &mdash; Hail</title>
  

  
  
    <link rel="shortcut icon" href="hail_logo_sq.png"/>
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="navbar.css" type="text/css" />
  
    <link rel="stylesheet" href="_static/rtd_modifications.css" type="text/css" />
  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Hail" href="index.html"/>
        <link rel="up" title="Other Resources" href="other_resources.html"/>
        <link rel="prev" title="Hadoop Glob Patterns" href="hadoop_glob_patterns.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>
  
  

</head>

<body class="wy-body-for-nav" role="document">

  <nav class="navbar navbar-default navbar-static-top" id="hail-navbar">
<div class="container-fluid" id="hail-container-fluid">
    <div class="navbar-header" id="hail-navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#hail-navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
   
      <a class="navbar-left" id="hail-navbar-brand" href="../index.html"><img alt="Hail" id="logo" src="hail-logo-cropped.png"></a>

    </div>

    <div class="collapse navbar-collapse" id="hail-navbar-collapse">
      <ul class="nav navbar-nav navbar-right">
          <li id="home" class="nav-item"><a href="../index.html">HOME</a></li>
          <li id="docs" class="nav-item"><a href="index.html">DOCS</a></li>
          <li id="forum" class="nav-item"><a href="http://discuss.hail.is">FORUM</a></li>
          <li id="chat" class="nav-item dropdown">
          <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">CHAT<span class="caret"></span></a>
          <ul class="dropdown-menu dropdown-menu-left">
            <li><a class="dropdown-item" href="https://gitter.im/hail-is/hail">General</a></li>
            <li><a class="dropdown-item" href="https://gitter.im/hail-is/hail-dev">Developers</a></li>
          </ul>
        </li>
       <li id="code" class="nav-item"><a href="https://github.com/hail-is/hail">CODE</a></li>
      </ul>
    </div>
  </div>
</nav>

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Hail
          

          
          </a>

          
            
            
              <div class="version">
                devel
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="exprlang.html">Expression Language</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">Python API</a></li>
<li class="toctree-l1"><a class="reference internal" href="annotationdb.html">Annotation Database</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="other_resources.html">Other Resources</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="hadoop_glob_patterns.html">Hadoop Glob Patterns</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">SQL</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#impala">Impala</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Hail</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="other_resources.html">Other Resources</a> &raquo;</li>
        
      <li>Querying using SQL</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/sql.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast,
.nboutput.nblast {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast + .nbinput {
    margin-top: -19px;
}

/* nice headers on first paragraph of info/warning boxes */
.admonition .first {
    margin: -12px;
    padding: 6px 12px;
    margin-bottom: 12px;
    color: #fff;
    line-height: 1;
    display: block;
}
.admonition.warning .first {
    background: #f0b37e;
}
.admonition.note .first {
    background: #6ab0de;
}
.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}
</style>
<div class="section" id="querying-using-sql">
<span id="sec-sql"></span><h1>Querying using SQL<a class="headerlink" href="#querying-using-sql" title="Permalink to this headline">¶</a></h1>
<p>Since Hail uses the Parquet file format for data storage, a Hail VDS can be queried using
Hadoop SQL tools, like Hive or Impala. This mode of access may be convenient for users
who have ad hoc queries that they are able to express in SQL.</p>
<p>Note that SQL access is <em>read-only</em>: it is not possible to write Hail datasets using
SQL at the current time.</p>
<div class="section" id="impala">
<h2>Impala<a class="headerlink" href="#impala" title="Permalink to this headline">¶</a></h2>
<p>Each VDS should be registered in the Hive metastore to allow Impala to query it (Impala uses Hive&#8217;s metastore to store table metadata). This is done by creating an external table in Hive, the &#8220;external&#8221; part means that the data is managed by an entity outside Hive (and
Impala). The table schema is read from one of the Parquet files in the VDS file
hierarchy.</p>
<p>To generate a Hive file:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Copy a VCF file into HDFS</li>
</ol>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ hadoop fs -put src/test/resources/sample.vcf.bgz sample.vcf.bgz
</pre></div>
</div>
<ol class="arabic" start="2">
<li><p class="first">Convert the VCF file into a VDS using Hail:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">hc</span><span class="o">.</span><span class="n">import_vcf</span><span class="p">(</span><span class="s2">&quot;sample.vcf.bgz&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;sample.vds&quot;</span><span class="p">,</span> <span class="n">parquet_genotypes</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>Note the use of <code class="docutils literal"><span class="pre">parquet_genotypes=True</span></code>, which writes the genotype
information using Parquet structures, rather than an opaque binary
representation that cannot be queried using SQL.</p>
</li>
<li><p class="first">Register the VDS as a Hive table</p>
</li>
</ol>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ PARQUET_DATA_FILE=$(hadoop fs -stat &#39;%n&#39; hdfs:///user/$USER/sample.vds/rdd.parquet/*.parquet | head -1)
$ impala-shell -q &quot;CREATE EXTERNAL TABLE variants LIKE PARQUET &#39;hdfs:///user/$USER/sample.vds/rdd.parquet/$PARQUET_DATA_FILE&#39; STORED AS PARQUET LOCATION &#39;hdfs:///user/$USER/sample.vds/rdd.parquet&#39;&quot;
</pre></div>
</div>
</div></blockquote>
<p>It is good practice to run Impala&#8217;s <code class="docutils literal"><span class="pre">COMPUTE</span> <span class="pre">STATS</span></code> command on the newly-created table, so that subsequent queries run efficiently.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ impala-shell -q &quot;COMPUTE STATS variants&quot;
</pre></div>
</div>
<p>Before running any queries it&#8217;s worth understanding the table schema, which is easily
done by calling <code class="docutils literal"><span class="pre">DESCRIBE</span></code> on the table:</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ impala-shell -q &quot;DESCRIBE variants&quot;
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>+-------------+----------------------------------+-----------------------------+
| name        | type                             | comment                     |
+-------------+----------------------------------+-----------------------------+
| variant     | struct&lt;                          | Inferred from Parquet file. |
|             |   contig:string,                 |                             |
|             |   start:int,                     |                             |
|             |   ref:string,                    |                             |
|             |   altalleles:array&lt;struct&lt;       |                             |
|             |     ref:string,                  |                             |
|             |     alt:string                   |                             |
|             |   &gt;&gt;                             |                             |
|             | &gt;                                |                             |
| annotations | struct&lt;                          | Inferred from Parquet file. |
|             |   rsid:string,                   |                             |
|             |   qual:double,                   |                             |
|             |   filters:array&lt;string&gt;,         |                             |
|             |   pass:boolean,                  |                             |
|             |   info:struct&lt;                   |                             |
|             |     negative_train_site:boolean, |                             |
|             |     hwp:double,                  |                             |
|             |     ac:array&lt;int&gt;,               |                             |
...
|             |   &gt;                              |                             |
|             | &gt;                                |                             |
| gs          | array&lt;struct&lt;                    | Inferred from Parquet file. |
|             |   gt:int,                        |                             |
|             |   ad:array&lt;int&gt;,                 |                             |
|             |   dp:int,                        |                             |
|             |   gq:int,                        |                             |
|             |   px:array&lt;int&gt;,                 |                             |
|             |   fakeref:boolean,               |                             |
|             |   isdosage:boolean               |                             |
|             | &gt;&gt;                               |                             |
+-------------+----------------------------------+-----------------------------+
</pre></div>
</div>
<p>Notice that the schema is nested. The <code class="docutils literal"><span class="pre">annotations</span></code> type corresponds to the variant
annotation schema that is returned by <a class="reference internal" href="hail.VariantDataset.html#hail.VariantDataset.variant_schema" title="hail.VariantDataset.variant_schema"><code class="xref py py-meth docutils literal"><span class="pre">variant_schema()</span></code></a>:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">print</span><span class="p">(</span><span class="n">hc</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s2">&quot;sample.vds&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">variant_schema</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>Struct {
    rsid: String,
    qual: Double,
    filters: Set[String],
    pass: Boolean,
    info: Struct {
        NEGATIVE_TRAIN_SITE: Boolean,
        HWP: Double,
        AC: Array[Int],
        ...
    }
}
</pre></div>
</div>
<p>Here is an example query to find variants in a given interval. Note the way that the
array of alternate alleles is joined with the main table, and the use of the
<code class="docutils literal"><span class="pre">item</span></code> keyword to refer to the value of the array element. Working with complex types
is explained in detail in the <a class="reference external" href="http://www.cloudera.com/documentation/enterprise/5-5-x/topics/impala_complex_types.html">Impala documentation</a>.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ impala-shell -q &quot;SELECT variant.contig, variant.start, variant.ref, altalleles.item.alt, annotations.rsid FROM variants, variants.variant.altalleles WHERE variant.start &gt; 13090000 AND variant.start &lt; 13100000&quot;
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>+----------------+---------------+-------------+----------+------------------+
| variant.contig | variant.start | variant.ref | item.alt | annotations.rsid |
+----------------+---------------+-------------+----------+------------------+
| 20             | 13090728      | A           | T        | rs6109712        |
| 20             | 13090733      | A           | AT       | .                |
| 20             | 13090733      | AT          | A        | .                |
| 20             | 13090745      | G           | C        | rs2236126        |
| 20             | 13098135      | T           | C        | rs150175260      |
+----------------+---------------+-------------+----------+------------------+
</pre></div>
</div>
<p>Here is another example showing how you can query the genotype information. Notice that
each genotype is represented by a whole row in the results. The <code class="docutils literal"><span class="pre">genotype_pos</span></code> column is
the index of the genotype for the variant.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ impala-shell -q &quot;SELECT variant.contig, variant.start, variant.ref, gs.pos AS genotype_pos, gs.item.gt AS gt FROM variants, variants.gs WHERE variant.start = 13090728 AND gs.pos &gt;= 20 AND gs.pos &lt; 25;&quot;
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>+----------------+---------------+-------------+--------------+----+
| variant.contig | variant.start | variant.ref | genotype_pos | gt |
+----------------+---------------+-------------+--------------+----+
| 20             | 13090728      | A           | 20           | 1  |
| 20             | 13090728      | A           | 21           | 0  |
| 20             | 13090728      | A           | 22           | 0  |
| 20             | 13090728      | A           | 23           | 0  |
| 20             | 13090728      | A           | 24           | 0  |
+----------------+---------------+-------------+--------------+----+
</pre></div>
</div>
<p>We can also retrieve the values from the AD (Allelic Depths) array by doing a nested
query that returns one row per genotype and per AD value. The <code class="docutils literal"><span class="pre">ad_pos</span></code> column is
the index of the value in the AD array.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ impala-shell -q &quot;SELECT variant.contig, variant.start, variant.ref, gs.pos AS genotype_pos, gs.item.gt AS gt, ad.pos AS ad_pos, ad.item AS ad FROM variants, variants.gs, gs.ad WHERE variant.start = 13090728 LIMIT 6;&quot;
</pre></div>
</div>
<div class="highlight-text"><div class="highlight"><pre><span></span>+----------------+---------------+-------------+--------------+----+--------+----+
| variant.contig | variant.start | variant.ref | genotype_pos | gt | ad_pos | ad |
+----------------+---------------+-------------+--------------+----+--------+----+
| 20             | 13090728      | A           | 0            | 0  | 0      | 28 |
| 20             | 13090728      | A           | 0            | 0  | 1      | 0  |
| 20             | 13090728      | A           | 1            | 0  | 0      | 20 |
| 20             | 13090728      | A           | 1            | 0  | 1      | 0  |
| 20             | 13090728      | A           | 2            | 0  | 0      | 11 |
| 20             | 13090728      | A           | 2            | 0  | 1      | 0  |
+----------------+---------------+-------------+--------------+----+--------+----+
</pre></div>
</div>
<p>If you no longer need to use SQL you can delete the table definition. Since the table
was registered as an external table the underlying data is <em>not</em> affected, so you can
still access the VDS from Hail.</p>
<div class="highlight-text"><div class="highlight"><pre><span></span>$ impala-shell -q &quot;DROP TABLE variants&quot;
$ hadoop fs -ls sample.vds
</pre></div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="hadoop_glob_patterns.html" class="btn btn-neutral" title="Hadoop Glob Patterns" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Hail Team.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'devel',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      <script type="text/javascript" src="../bootstrap.min.js"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
  
  
   

</body>
</html>