BEGIN TRANSACTION;
CREATE TABLE "files" (
	`file_id`	TEXT,
	`file_path`	TEXT,
	`file_type`	TEXT,
	`key_type`	TEXT
);
INSERT INTO `files` (file_id,file_path,file_type,key_type) VALUES ('f_0000','gs://annotationdb/cadd/cadd.vds','vds','variant'),
 ('f_0001','gs://annotationdb/chromHMM/enhancers.kt','table','interval'),
 ('f_0003','gs://annotationdb/dann/dann.vds','vds','variant'),
 ('f_0004','gs://annotationdb/dbNSFP/dbNSFP.vds','vds','variant'),
 ('f_0005','gs://annotationdb/dhs/dhs.kt','table','interval'),
 ('f_0007','gs://annotationdb/eigen/eigen.vds','vds','variant'),
 ('f_0008','gs://annotationdb/fantom5/robust_enhancers.kt','table','interval'),
 ('f_0009','gs://annotationdb/funseq2/funseq2.vds','vds','variant'),
 ('f_0010','gs://annotationdb/gencode19/gencode19.kt','table','interval'),
 ('f_0011','gs://annotationdb/gene/gene.kt','table','gene'),
 ('f_0002','gs://annotationdb/chromHMM/promoters.kt','table','interval'),
 ('f_0006','gs://annotationdb/discovEHR/discovEHR.vds','vds','variant'),
 ('f_0013','gs://annotationdb/gnomAD/exomes.vds','vds','variant'),
 ('f_0014','gs://annotationdb/gnomAD/genomes.vds','vds','variant'),
 ('f_0015','gs://annotationdb/isLCR/isLCR.kt','table','interval'),
 ('f_0016','gs://annotationdb/ldscore/ldscore.kt','table','interval'),
 ('f_0017','gs://annotationdb/linsight/linsight.kt','table','interval'),
 ('f_0018','gs://annotationdb/segway/segway.kt','table','interval'),
 ('f_0019','gs://annotationdb/tfbs/tfbs.kt','table','variant'),
 ('f_0020','gs://annotationdb/vista/vista.kt','table','interval'),
 ('vep','vep','vds','variant'),
 ('f_0021','gs://annotationdb/fantom5/permissive_enhancers.kt','table','interval');
CREATE TABLE "docs" (
	`parent`	TEXT,
	`annotation`	TEXT,
	`text`	TEXT,
	`study_link`	TEXT,
	`study_data`	TEXT,
	`study_title`	TEXT,
	`free_text`	TEXT,
	`selectable`	INTEGER,
	PRIMARY KEY(`annotation`)
);
INSERT INTO `docs` (parent,annotation,text,study_link,study_data,study_title,free_text,selectable) VALUES ('va','va.cadd','CADD','https://www.ncbi.nlm.nih.gov/pubmed/24487276','http://cadd.gs.washington.edu/download','Kircher, Martin et al. "A General Framework for Estimating the Relative Pathogenicity of Human Genetic Variants."Nature genetics 46.3 (2014): 310-315.','This is CADD version 1.3. For more info, we suggest reading through the official release notes: http://cadd.gs.washington.edu/static/ReleaseNotes_CADD_v1.3.pdf',0),
 ('va','va.chromHMM','Epigenome Roadmap - ChromHMM','https://www.ncbi.nlm.nih.gov/pubmed/?term=25693563','https://personal.broadinstitute.org/meuleman/reg2map/HoneyBadger2-impute_release/','Roadmap Epigenomics Consortium et al. “Integrative Analysis of 111 Reference Human Epigenomes.” Nature 518.7539 (2015): 317–330.','Chromatin state discovery and characterization. 

Data were downloaded in R format. We included only intervals overlapping with DNaseI regions selected with -log10(p) >= 10. 

Information about the interpretation of each of the 25 states can be found here: http://egg2.wustl.edu/roadmap/web_portal/imputed.html#chr_imp. 

The name for each cell type can be found here: https://docs.google.com/spreadsheets/u/1/d/1yikGx4MsO9Ei36b64yOy9Vb6oPC5IBGlFbYEt-N6gOM/edit?usp=sharing#gid=15.',0),
 ('va.chromHMM','va.chromHMM.enhancers','Enhancers',NULL,NULL,NULL,'Enhancers states included: 10, 11, 12, 13, 14, 15, 16, 17, 18. Strong enhancers usually defined: 10, 12, 13, 14, 15. 0=no enhancer.',1),
 ('va.chromHMM','va.chromHMM.promoters','Promoters',NULL,NULL,NULL,'Promoters + TSS states included: 1, 2, 3, 4, 22, 23. Strong promoters + TSS usually defined: 1, 2, 3, 4. 1=TSS, 0=no promoters, TSS.',1),
 ('va','va.dann','DANN','https://www.ncbi.nlm.nih.gov/pubmed/25338716','https://cbcl.ics.uci.edu/public_data/DANN/data/','Quang, Daniel, Yifei Chen, and Xiaohui Xie. "DANN: A Deep Learning Approach for Annotating the Pathogenicity of Genetic Variants." Bioinformatics 31.5 (2015): 761-763.','Same as CADD, but using a deep learning approach.',0),
 ('va','va.dbNSFP','dbNSFP v2.9.3','https://www.ncbi.nlm.nih.gov/pubmed/?term=26555599','ftp://dbnsfp:dbnsfp@dbnsfp.softgenetics.com/dbNSFPv2.9.3.zip','Liu, Xiaoming et al. "dbNSFP v3.0: A One-Stop Database of Functional Predictions and Annotations for Human Non-Synonymous and Splice Site SNVs." Human mutation 37.3 (2016): 235-241.','Database for Nonsynonymous SNPs'' functional predictions, v2.9.3.

The dbNSFP is an integrated database of functional annotations from multiple sources for the comprehensive collection of human non-synonymous SNPs (NSs).

Its current version includes a total of 87,361,054 NSs. It compiles prediction scores from 14 prediction algorithms (SIFT, Polyphen2, LRT, MutationTaster, MutationAssessor, FATHMM, MetaSVM, MetaLR, CADD, VEST, PROVEAN, M-CAP, Eigen\uff0cEigen PC), eight conservation scores (phyloP46way_primate, phyloP46way_placental, phyloP100way_vertebrate, phastCons46way_primate, phastCons46way_placental, phastCons100way_veterbrate, GERP++, and SiPhy) and other function annotations.

We included all the original annotations except for allele frequencies in 1000 genome and ExAC since these are covered in other annotations.

Note 1: Missing data is designated as ".".
Note 2: The three Uniprot entries are based on Polyphen-2 v2.2.2 annotations.',0),
 ('va','va.dhs','Replicated DHS','http://biorxiv.org/content/early/2016/05/19/054361','Shared by pshoosh@broadinstitute.org','Shooshtari et Al. Integrative genetic and epigenetic analysis uncovers regulatory mechanisms of autoimmune disease','Detailed information about the processing are available https://storage.googleapis.com/annotationdb/dhs/Replicable-DHS.pdf',0),
 ('va','va.discovEHR','DiscovEHR','https://www.ncbi.nlm.nih.gov/pubmed/28008009','http://www.discovehrshare.com/downloads','Distribution and clinical impact of functional variants in 50,726 whole-exome sequences from the DiscovEHR study. Science (2016).',NULL,0),
 ('va','va.eigen','EIGEN','https://www.ncbi.nlm.nih.gov/pubmed/26727659','https://xioniti01.u.hpc.mssm.edu/v1.1/','IONITA-LAZA, IULIANA et al. "A SPECTRAL APPROACH INTEGRATING FUNCTIONAL GENOMIC ANNOTATIONS FOR CODING AND NONCODING VARIANTS." Nature genetics 48.2 (2016): 214-220.','EIGEN v1.1. In the paper, Eigen PC works better than EIGEN in non-coding regions.',0),
 ('va','va.fantom5','FANTOM 5','https://www.ncbi.nlm.nih.gov/pubmed/?term=24670763','http://enhancer.binf.ku.dk/presets/robust_enhancers.bed and http://enhancer.binf.ku.dk/presets/permissive_enhancers.bed','Andersson, Robin et al. “An Atlas of Active Enhancers across Human Cell Types and Tissues.” Nature 507.7493 (2014): 455–461.','FANTOM5 enhancers obtained with CAGE technique',0),
 ('va','va.funseq2','FUNSEQ 2','https://www.ncbi.nlm.nih.gov/pubmed/25273974','http://archive.gersteinlab.org/funseq2.1.2/hg19_NCscore_funseq216.tsv.bgz','Fu, Yao et al. “FunSeq2: A Framework for Prioritizing Noncoding Regulatory Variants in Cancer.” Genome Biology 15.10 (2014): 480. PMC. Web. 19 June 2017.','FUNSEQ 2 is an approach developed for prioritizing somatic cancer mutations.',0),
 ('va','va.gencode19','GENCODE v19',NULL,NULL,NULL,'These files were obtained from UCSC table browser with these settings: Assembly: Feb. 2009 (GRCh37/hg19), Group: Genes and Gene Predictions, Track: GENCODE Genes V19, Region:  genome, Output: format: BED',0),
 ('va','va.gene','Gene-level annotations',NULL,NULL,NULL,'These gene-level annotations can be used to annotate variants in a VDS by specifying a single gene symbol for each variant. 

This gene key is used to map each variant to the appropriate gene-level annotations.

By default, VEP is run and its output is parsed to extract the gene key (see more detailed description on the bottom of this page).',0),
 ('va.gene','va.gene.names','Gene names','https://www.ncbi.nlm.nih.gov/pubmed/?term=26555599','ftp://dbnsfp:dbnsfp@dbnsfp.softgenetics.com/dbNSFPv2.9.3.zip','Liu, Xiaoming et al. "dbNSFP v3.0: A One-Stop Database of Functional Predictions and Annotations for Human Non-Synonymous and Splice Site SNVs."','Part of the dbnsfp database contains gene-level annotations. The original file is called dbNSFP2.9_gene. We include here few annotations that allow to exchange gene names across databases.',1),
 ('va.gene','va.gene.constraint','Gene constraint scores','https://www.ncbi.nlm.nih.gov/pubmed/27535533','ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3.1/functional_gene_constraint','Exome Aggregation Consortium et al. “Analysis of Protein-Coding Genetic Variation in 60,706 Humans.” Nature 536.7616 (2016): 285–291.','Detailed description of the content of these scores is provided at: ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3.1/functional_gene_constraint. 

No TGCA: exclude TGCA samples, No psych: exclude samples with psychiatric disorders',1),
 ('va','va.gnomAD','gnomAD','https://www.ncbi.nlm.nih.gov/pubmed/27535533','http://gnomad.broadinstitute.org/downloads','Exome Aggregation Consortium et al. "Analysis of Protein-Coding Genetic Variation in 60,706 Humans." Nature 536.7616 (2016): 285-291.','The Genome Aggregation Database (gnomAD) is a resource developed by an international coalition of investigators, with the goal of aggregating and harmonizing both exome and genome sequencing data from a wide variety of large-scale sequencing projects, and making summary data available for the wider scientific community.

The data set provided on this website spans 123,136 exome sequences and 15,496 whole-genome sequences from unrelated individuals sequenced as part of various disease-specific and population genetic studies. 

The gnomAD Principal Investigators and groups that have contributed data to the current release are listed <a href=\"http://gnomad.broadinstitute.org/about\" target=\"_blank\">here</a>.\n\nAll data here are released for the benefit of the wider biomedical community, without restriction on use - see the terms of use <a href=\"http://gnomad.broadinstitute.org/terms\" target=\"_blank\">here</a>.',0),
 ('va.gnomAD','va.gnomAD.exomes','Exomes',NULL,NULL,NULL,'Annotations generated from 123,136 whole exome sequences in gnomAD.',1),
 ('va.gnomAD','va.gnomAD.genomes','Genomes',NULL,NULL,NULL,'Annotations generated from 15,496 whole genome sequences in gnomAD.',1),
 ('va','va.isLCR','Low-complexity regions','https://www.ncbi.nlm.nih.gov/pubmed/?term=24974202','https://github.com/lh3/varcmp/blob/master/scripts/LCR-hs37d5.bed.gz','Li, Heng. “Toward Better Understanding of Artifacts in Variant Calling from High-Coverage Samples.” Bioinformatics 30.20 (2014): 2843–2851.','Regions with repeated sequences normally excluded from WGS studies',1),
 ('va','va.ldscore','LD SCORE Annotations','https://www.ncbi.nlm.nih.gov/pubmed/?term=26414678','https://data.broadinstitute.org/alkesgroup/LDSCORE/baseline_bedfiles.tgz','Finucane, Hilary K. et al. “Partitioning Heritability by Functional Annotation Using Genome-Wide Association Summary Statistics.” Nature genetics 47.11 (2015): 1228–1235. PMC. Web. 19 June 2017.','Set of annotations used for LDSCORE partitioning heritability. Info for each annotation are available in the original publication.',0),
 ('va','va.linsight','LINSIGHT','http://www.nature.com/ng/journal/v49/n4/full/ng.3810.html','http://compgen.cshl.edu/~yihuang/tracks/LINSIGHT.bw','Huang, Yi-Fei, Brad Gulko, and Adam Siepel. "Fast, scalable prediction of deleterious noncoding variants from functional and population genomic data" Nature Genetics. Online.','We tranformed the .bw data into .bed format using this program: http://hgdownload.soe.ucsc.edu/admin/exe/linux.x86_64/bigWigToBedGraph',0),
 ('va','va.segway','SEGWAY','http://biorxiv.org/content/early/2016/11/07/086025','http://noble.gs.washington.edu/proj/encyclopedia/segway_encyclopedia.bed.gz','A unified encyclopedia of human functional elements through fully automated annotation of 164 human cell types. Biorxiv.','Data were imported as it is from the website. The sum of functionality score, the average base-wise functionality score and the majority label of each cell type in the segment are reported. Information about the interpretation of each label can be found here: http://noble.gs.washington.edu/proj/encyclopedia/',0),
 ('va','va.tfbs','Transcription factor binding sites',NULL,'http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeeRegTfbsClusteredV3.bed.gz',NULL,'Transcription factor binding sites obtained from V3 of ENCODE and downloaded from UCSC browser. The track is described here: http://genome.ucsc.edu/cgi-bin/hgTrackUi?db=hg19&g=wgEncodeRegTfbsClusteredV3',0),
 ('va','va.vep','Variant Effect Predictor (VEP)','https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4893825/','http://grch37.ensembl.org/Homo_sapiens/Tools/VEP','McLaren, William et al. "The Ensembl Variant Effect Predictor." Genome Biology 17 (2016): 122.','Variant Effect Predictor (VEP) with Loss-Of-Function Transcript Effect Estimator (LOFTEE) plug-in.',1),
 ('va','va.vista','VISTA','https://www.ncbi.nlm.nih.gov/pubmed/?term=17130149','https://enhancer.lbl.gov/cgi-bin/imagedb3.pl?page_size=20000;show=1;search.result=yes;form=search;search.form=no;action=search;search.sequence=1','Visel, Axel et al. “VISTA Enhancer Browser—a Database of Tissue-Specific Human Enhancers.” Nucleic Acids Research 35.Database issue (2007): D88–D92.','Validated enhancers from VISTA. We select only positive human elements. 

Once downloaded, 2APR2017, the data were processed as follows: 

''grep \">Human\" vista_2APR2017.txt | grep \"positive\" | awk ''{split($0,a,\"|\"); print a[2], a[5]}'' > vista_2APR2017_processed.txt''

More info available here: https://enhancer.lbl.gov/aboutproject_n.html.',0),
 ('','va','','','','','','');
CREATE TABLE `array_elements` (
	`annotation`	TEXT,
	`element_name`	TEXT,
	`element_type`	TEXT
);
INSERT INTO `array_elements` (annotation,element_name,element_type) VALUES ('va.vep.colocated_variants','aa_allele','String'),
 ('va.vep.colocated_variants','aa_maf','Double'),
 ('va.vep.colocated_variants','afr_allele','String'),
 ('va.vep.colocated_variants','afr_maf','Double'),
 ('va.vep.colocated_variants','allele_string','String'),
 ('va.vep.colocated_variants','amr_allele','String'),
 ('va.vep.colocated_variants','amr_maf','Double'),
 ('va.vep.colocated_variants','clin_sig','Array[String]'),
 ('va.vep.colocated_variants','end','Int'),
 ('va.vep.colocated_variants','eas_allele','String'),
 ('va.vep.colocated_variants','eas_maf','Double'),
 ('va.vep.colocated_variants','ea_allele','String'),
 ('va.vep.colocated_variants','ea_maf','Double'),
 ('va.vep.colocated_variants','eur_allele','String'),
 ('va.vep.colocated_variants','eur_maf','Double'),
 ('va.vep.colocated_variants','exac_adj_allele','String'),
 ('va.vep.colocated_variants','exac_adj_maf','Double'),
 ('va.vep.colocated_variants','exac_allele','String'),
 ('va.vep.colocated_variants','exac_afr_allele','String'),
 ('va.vep.colocated_variants','exac_afr_maf','Double'),
 ('va.vep.colocated_variants','exac_amr_allele','String'),
 ('va.vep.colocated_variants','exac_amr_maf','Double'),
 ('va.vep.colocated_variants','exac_eas_allele','String'),
 ('va.vep.colocated_variants','exac_eas_maf','Double'),
 ('va.vep.colocated_variants','exac_fin_allele','String'),
 ('va.vep.colocated_variants','exac_fin_maf','Double'),
 ('va.vep.colocated_variants','exac_maf','Double'),
 ('va.vep.colocated_variants','exac_nfe_allele','String'),
 ('va.vep.colocated_variants','exac_nfe_maf','Double'),
 ('va.vep.colocated_variants','exac_oth_allele','String'),
 ('va.vep.colocated_variants','exac_oth_maf','Double'),
 ('va.vep.colocated_variants','exac_sas_allele','String'),
 ('va.vep.colocated_variants','exac_sas_maf','Double'),
 ('va.vep.colocated_variants','id','String'),
 ('va.vep.colocated_variants','minor_allele','String'),
 ('va.vep.colocated_variants','minor_allele_freq','Double'),
 ('va.vep.colocated_variants','phenotype_or_disease','Int'),
 ('va.vep.colocated_variants','pubmed','Array[Int]'),
 ('va.vep.colocated_variants','sas_allele','String'),
 ('va.vep.colocated_variants','sas_maf','Double'),
 ('va.vep.colocated_variants','somatic','Int'),
 ('va.vep.colocated_variants','start','Int'),
 ('va.vep.colocated_variants','strand','Int'),
 ('va.vep.intergenic_consequences','allele_num','Int'),
 ('va.vep.intergenic_consequences','consequence_terms','Array[String]'),
 ('va.vep.intergenic_consequences','impact','String'),
 ('va.vep.intergenic_consequences','minimised','Int'),
 ('va.vep.intergenic_consequences','variant_allele','String'),
 ('va.vep.motif_feature_consequences','allele_num','Int'),
 ('va.vep.motif_feature_consequences','consequence_terms','Array[String]'),
 ('va.vep.motif_feature_consequences','high_inf_pos','String'),
 ('va.vep.motif_feature_consequences','impact','String'),
 ('va.vep.motif_feature_consequences','minimised','Int'),
 ('va.vep.motif_feature_consequences','motif_feature_id','String'),
 ('va.vep.motif_feature_consequences','motif_name','String'),
 ('va.vep.motif_feature_consequences','motif_pos','Int'),
 ('va.vep.motif_feature_consequences','motif_score_change','Double'),
 ('va.vep.motif_feature_consequences','strand','Int'),
 ('va.vep.motif_feature_consequences','variant_allele','String'),
 ('va.vep.regulatory_feature_consequences','allele_num','Int'),
 ('va.vep.regulatory_feature_consequences','biotype','String'),
 ('va.vep.regulatory_feature_consequences','consequence_terms','Array[String]'),
 ('va.vep.regulatory_feature_consequences','impact','String'),
 ('va.vep.regulatory_feature_consequences','minimised','Int'),
 ('va.vep.regulatory_feature_consequences','regulatory_feature_id','String'),
 ('va.vep.regulatory_feature_consequences','variant_allele','String'),
 ('va.vep.transcript_consequences','allele_num','Int'),
 ('va.vep.transcript_consequences','amino_acids','String'),
 ('va.vep.transcript_consequences','biotype','String'),
 ('va.vep.transcript_consequences','canonical','Int'),
 ('va.vep.transcript_consequences','ccds','String'),
 ('va.vep.transcript_consequences','cdna_start','Int'),
 ('va.vep.transcript_consequences','cdna_end','Int'),
 ('va.vep.transcript_consequences','cds_end','Int'),
 ('va.vep.transcript_consequences','cds_start','Int'),
 ('va.vep.transcript_consequences','codons','String'),
 ('va.vep.transcript_consequences','consequence_terms','Array[String]'),
 ('va.vep.transcript_consequences','distance','Int'),
 ('va.vep.transcript_consequences','domains','Array[Struct]'),
 ('va.vep.transcript_consequences','exon','String'),
 ('va.vep.transcript_consequences','gene_id','String'),
 ('va.vep.transcript_consequences','gene_pheno','Int'),
 ('va.vep.transcript_consequences','gene_symbol','String'),
 ('va.vep.transcript_consequences','gene_symbol_source','String'),
 ('va.vep.transcript_consequences','hgnc_id','String'),
 ('va.vep.transcript_consequences','hgvsc','String'),
 ('va.vep.transcript_consequences','hgvsp','String'),
 ('va.vep.transcript_consequences','hgvs_offset','Int'),
 ('va.vep.transcript_consequences','impact','String'),
 ('va.vep.transcript_consequences','intron','String'),
 ('va.vep.transcript_consequences','lof','String'),
 ('va.vep.transcript_consequences','lof_flags','String'),
 ('va.vep.transcript_consequences','lof_filter','String'),
 ('va.vep.transcript_consequences','lof_info','String'),
 ('va.vep.transcript_consequences','minimised','Int'),
 ('va.vep.transcript_consequences','polyphen_prediction','String'),
 ('va.vep.transcript_consequences','polyphen_score','Double'),
 ('va.vep.transcript_consequences','protein_end','Int'),
 ('va.vep.transcript_consequences','protein_start','Int'),
 ('va.vep.transcript_consequences','protein_id','String'),
 ('va.vep.transcript_consequences','sift_prediction','String'),
 ('va.vep.transcript_consequences','sift_score','Double'),
 ('va.vep.transcript_consequences','strand','Int'),
 ('va.vep.transcript_consequences','transcript_id','String'),
 ('va.vep.transcript_consequences','trembl','String'),
 ('va.vep.transcript_consequences','uniparc','String'),
 ('va.vep.transcript_consequences','variant_allele','String');
CREATE TABLE "annotations" (
	`file_id`	TEXT,
	`parent`	TEXT,
	`annotation`	TEXT,
	`file_element`	TEXT,
	`type`	TEXT,
	`description`	TEXT,
	PRIMARY KEY(`annotation`)
);
INSERT INTO `annotations` (file_id,parent,annotation,file_element,type,description) VALUES ('f_0003','va.dann','va.dann.score','score','Double','DANN score'),
 ('f_0000','va.cadd','va.cadd.RawScore','RawScore','Double','CADD raw score'),
 ('f_0000','va.cadd','va.cadd.PHRED','PHRED','Double','CADD PHRED-scaled score'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E001','E001','Int','Enhancers - Cell Type: E001'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E002','E002','Int','Enhancers - Cell Type: E002'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E003','E003','Int','Enhancers - Cell Type: E003'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E004','E004','Int','Enhancers - Cell Type: E004'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E005','E005','Int','Enhancers - Cell Type: E005'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E006','E006','Int','Enhancers - Cell Type: E006'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E007','E007','Int','Enhancers - Cell Type: E007'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E008','E008','Int','Enhancers - Cell Type: E008'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E009','E009','Int','Enhancers - Cell Type: E009'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E010','E010','Int','Enhancers - Cell Type: E010'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E011','E011','Int','Enhancers - Cell Type: E011'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E012','E012','Int','Enhancers - Cell Type: E012'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E013','E013','Int','Enhancers - Cell Type: E013'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E014','E014','Int','Enhancers - Cell Type: E014'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E015','E015','Int','Enhancers - Cell Type: E015'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E016','E016','Int','Enhancers - Cell Type: E016'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E017','E017','Int','Enhancers - Cell Type: E017'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E018','E018','Int','Enhancers - Cell Type: E018'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E019','E019','Int','Enhancers - Cell Type: E019'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E020','E020','Int','Enhancers - Cell Type: E020'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E021','E021','Int','Enhancers - Cell Type: E021'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E022','E022','Int','Enhancers - Cell Type: E022'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E023','E023','Int','Enhancers - Cell Type: E023'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E024','E024','Int','Enhancers - Cell Type: E024'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E025','E025','Int','Enhancers - Cell Type: E025'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E026','E026','Int','Enhancers - Cell Type: E026'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E027','E027','Int','Enhancers - Cell Type: E027'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E028','E028','Int','Enhancers - Cell Type: E028'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E029','E029','Int','Enhancers - Cell Type: E029'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E030','E030','Int','Enhancers - Cell Type: E030'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E031','E031','Int','Enhancers - Cell Type: E031'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E032','E032','Int','Enhancers - Cell Type: E032'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E033','E033','Int','Enhancers - Cell Type: E033'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E034','E034','Int','Enhancers - Cell Type: E034'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E035','E035','Int','Enhancers - Cell Type: E035'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E036','E036','Int','Enhancers - Cell Type: E036'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E037','E037','Int','Enhancers - Cell Type: E037'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E038','E038','Int','Enhancers - Cell Type: E038'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E039','E039','Int','Enhancers - Cell Type: E039'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E040','E040','Int','Enhancers - Cell Type: E040'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E041','E041','Int','Enhancers - Cell Type: E041'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E042','E042','Int','Enhancers - Cell Type: E042'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E043','E043','Int','Enhancers - Cell Type: E043'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E044','E044','Int','Enhancers - Cell Type: E044'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E045','E045','Int','Enhancers - Cell Type: E045'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E046','E046','Int','Enhancers - Cell Type: E046'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E047','E047','Int','Enhancers - Cell Type: E047'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E048','E048','Int','Enhancers - Cell Type: E048'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E049','E049','Int','Enhancers - Cell Type: E049'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E050','E050','Int','Enhancers - Cell Type: E050'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E051','E051','Int','Enhancers - Cell Type: E051'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E052','E052','Int','Enhancers - Cell Type: E052'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E053','E053','Int','Enhancers - Cell Type: E053'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E054','E054','Int','Enhancers - Cell Type: E054'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E055','E055','Int','Enhancers - Cell Type: E055'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E056','E056','Int','Enhancers - Cell Type: E056'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E057','E057','Int','Enhancers - Cell Type: E057'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E058','E058','Int','Enhancers - Cell Type: E058'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E059','E059','Int','Enhancers - Cell Type: E059'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E061','E061','Int','Enhancers - Cell Type: E061'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E062','E062','Int','Enhancers - Cell Type: E062'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E063','E063','Int','Enhancers - Cell Type: E063'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E065','E065','Int','Enhancers - Cell Type: E065'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E066','E066','Int','Enhancers - Cell Type: E066'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E067','E067','Int','Enhancers - Cell Type: E067'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E068','E068','Int','Enhancers - Cell Type: E068'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E069','E069','Int','Enhancers - Cell Type: E069'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E070','E070','Int','Enhancers - Cell Type: E070'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E071','E071','Int','Enhancers - Cell Type: E071'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E072','E072','Int','Enhancers - Cell Type: E072'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E073','E073','Int','Enhancers - Cell Type: E073'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E074','E074','Int','Enhancers - Cell Type: E074'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E075','E075','Int','Enhancers - Cell Type: E075'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E076','E076','Int','Enhancers - Cell Type: E076'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E077','E077','Int','Enhancers - Cell Type: E077'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E078','E078','Int','Enhancers - Cell Type: E078'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E079','E079','Int','Enhancers - Cell Type: E079'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E080','E080','Int','Enhancers - Cell Type: E080'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E081','E081','Int','Enhancers - Cell Type: E081'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E082','E082','Int','Enhancers - Cell Type: E082'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E083','E083','Int','Enhancers - Cell Type: E083'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E084','E084','Int','Enhancers - Cell Type: E084'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E085','E085','Int','Enhancers - Cell Type: E085'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E086','E086','Int','Enhancers - Cell Type: E086'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E087','E087','Int','Enhancers - Cell Type: E087'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E088','E088','Int','Enhancers - Cell Type: E088'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E089','E089','Int','Enhancers - Cell Type: E089'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E090','E090','Int','Enhancers - Cell Type: E090'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E091','E091','Int','Enhancers - Cell Type: E091'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E092','E092','Int','Enhancers - Cell Type: E092'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E093','E093','Int','Enhancers - Cell Type: E093'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E094','E094','Int','Enhancers - Cell Type: E094'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E095','E095','Int','Enhancers - Cell Type: E095'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E096','E096','Int','Enhancers - Cell Type: E096'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E097','E097','Int','Enhancers - Cell Type: E097'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E098','E098','Int','Enhancers - Cell Type: E098'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E099','E099','Int','Enhancers - Cell Type: E099'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E100','E100','Int','Enhancers - Cell Type: E100'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E101','E101','Int','Enhancers - Cell Type: E101'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E102','E102','Int','Enhancers - Cell Type: E102'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E103','E103','Int','Enhancers - Cell Type: E103'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E104','E104','Int','Enhancers - Cell Type: E104'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E105','E105','Int','Enhancers - Cell Type: E105'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E106','E106','Int','Enhancers - Cell Type: E106'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E107','E107','Int','Enhancers - Cell Type: E107'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E108','E108','Int','Enhancers - Cell Type: E108'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E109','E109','Int','Enhancers - Cell Type: E109'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E110','E110','Int','Enhancers - Cell Type: E110'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E111','E111','Int','Enhancers - Cell Type: E111'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E112','E112','Int','Enhancers - Cell Type: E112'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E113','E113','Int','Enhancers - Cell Type: E113'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E114','E114','Int','Enhancers - Cell Type: E114'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E115','E115','Int','Enhancers - Cell Type: E115'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E116','E116','Int','Enhancers - Cell Type: E116'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E117','E117','Int','Enhancers - Cell Type: E117'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E118','E118','Int','Enhancers - Cell Type: E118'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E119','E119','Int','Enhancers - Cell Type: E119'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E120','E120','Int','Enhancers - Cell Type: E120'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E121','E121','Int','Enhancers - Cell Type: E121'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E122','E122','Int','Enhancers - Cell Type: E122'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E123','E123','Int','Enhancers - Cell Type: E123'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E124','E124','Int','Enhancers - Cell Type: E124'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E125','E125','Int','Enhancers - Cell Type: E125'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E126','E126','Int','Enhancers - Cell Type: E126'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E127','E127','Int','Enhancers - Cell Type: E127'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E001','E001','Int','Promoters - Cell Type: E001'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E002','E002','Int','Promoters - Cell Type: E002'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E003','E003','Int','Promoters - Cell Type: E003'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E004','E004','Int','Promoters - Cell Type: E004'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E005','E005','Int','Promoters - Cell Type: E005'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E006','E006','Int','Promoters - Cell Type: E006'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E007','E007','Int','Promoters - Cell Type: E007'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E008','E008','Int','Promoters - Cell Type: E008'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E009','E009','Int','Promoters - Cell Type: E009'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E010','E010','Int','Promoters - Cell Type: E010'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E011','E011','Int','Promoters - Cell Type: E011'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E012','E012','Int','Promoters - Cell Type: E012'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E013','E013','Int','Promoters - Cell Type: E013'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E014','E014','Int','Promoters - Cell Type: E014'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E015','E015','Int','Promoters - Cell Type: E015'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E016','E016','Int','Promoters - Cell Type: E016'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E017','E017','Int','Promoters - Cell Type: E017'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E018','E018','Int','Promoters - Cell Type: E018'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E019','E019','Int','Promoters - Cell Type: E019'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E020','E020','Int','Promoters - Cell Type: E020'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E021','E021','Int','Promoters - Cell Type: E021'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E022','E022','Int','Promoters - Cell Type: E022'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E023','E023','Int','Promoters - Cell Type: E023'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E024','E024','Int','Promoters - Cell Type: E024'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E025','E025','Int','Promoters - Cell Type: E025'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E026','E026','Int','Promoters - Cell Type: E026'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E027','E027','Int','Promoters - Cell Type: E027'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E028','E028','Int','Promoters - Cell Type: E028'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E029','E029','Int','Promoters - Cell Type: E029'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E030','E030','Int','Promoters - Cell Type: E030'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E031','E031','Int','Promoters - Cell Type: E031'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E032','E032','Int','Promoters - Cell Type: E032'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E033','E033','Int','Promoters - Cell Type: E033'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E034','E034','Int','Promoters - Cell Type: E034'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E035','E035','Int','Promoters - Cell Type: E035'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E036','E036','Int','Promoters - Cell Type: E036'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E037','E037','Int','Promoters - Cell Type: E037'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E038','E038','Int','Promoters - Cell Type: E038'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E039','E039','Int','Promoters - Cell Type: E039'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E040','E040','Int','Promoters - Cell Type: E040'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E041','E041','Int','Promoters - Cell Type: E041'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E042','E042','Int','Promoters - Cell Type: E042'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E043','E043','Int','Promoters - Cell Type: E043'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E044','E044','Int','Promoters - Cell Type: E044'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E045','E045','Int','Promoters - Cell Type: E045'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E046','E046','Int','Promoters - Cell Type: E046'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E047','E047','Int','Promoters - Cell Type: E047'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E048','E048','Int','Promoters - Cell Type: E048'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E049','E049','Int','Promoters - Cell Type: E049'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E050','E050','Int','Promoters - Cell Type: E050'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E051','E051','Int','Promoters - Cell Type: E051'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E052','E052','Int','Promoters - Cell Type: E052'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E053','E053','Int','Promoters - Cell Type: E053'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E054','E054','Int','Promoters - Cell Type: E054'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E055','E055','Int','Promoters - Cell Type: E055'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E056','E056','Int','Promoters - Cell Type: E056'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E057','E057','Int','Promoters - Cell Type: E057'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E058','E058','Int','Promoters - Cell Type: E058'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E059','E059','Int','Promoters - Cell Type: E059'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E061','E061','Int','Promoters - Cell Type: E061'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E062','E062','Int','Promoters - Cell Type: E062'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E063','E063','Int','Promoters - Cell Type: E063'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E065','E065','Int','Promoters - Cell Type: E065'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E066','E066','Int','Promoters - Cell Type: E066'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E067','E067','Int','Promoters - Cell Type: E067'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E068','E068','Int','Promoters - Cell Type: E068'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E069','E069','Int','Promoters - Cell Type: E069'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E070','E070','Int','Promoters - Cell Type: E070'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E071','E071','Int','Promoters - Cell Type: E071'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E072','E072','Int','Promoters - Cell Type: E072'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E073','E073','Int','Promoters - Cell Type: E073'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E074','E074','Int','Promoters - Cell Type: E074'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E075','E075','Int','Promoters - Cell Type: E075'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E076','E076','Int','Promoters - Cell Type: E076'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E077','E077','Int','Promoters - Cell Type: E077'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E078','E078','Int','Promoters - Cell Type: E078'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E079','E079','Int','Promoters - Cell Type: E079'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E080','E080','Int','Promoters - Cell Type: E080'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E081','E081','Int','Promoters - Cell Type: E081'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E082','E082','Int','Promoters - Cell Type: E082'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E083','E083','Int','Promoters - Cell Type: E083'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E084','E084','Int','Promoters - Cell Type: E084'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E085','E085','Int','Promoters - Cell Type: E085'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E086','E086','Int','Promoters - Cell Type: E086'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E087','E087','Int','Promoters - Cell Type: E087'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E088','E088','Int','Promoters - Cell Type: E088'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E089','E089','Int','Promoters - Cell Type: E089'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E090','E090','Int','Promoters - Cell Type: E090'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E091','E091','Int','Promoters - Cell Type: E091'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E092','E092','Int','Promoters - Cell Type: E092'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E093','E093','Int','Promoters - Cell Type: E093'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E094','E094','Int','Promoters - Cell Type: E094'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E095','E095','Int','Promoters - Cell Type: E095'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E096','E096','Int','Promoters - Cell Type: E096'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E097','E097','Int','Promoters - Cell Type: E097'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E098','E098','Int','Promoters - Cell Type: E098'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E099','E099','Int','Promoters - Cell Type: E099'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E100','E100','Int','Promoters - Cell Type: E100'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E101','E101','Int','Promoters - Cell Type: E101'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E102','E102','Int','Promoters - Cell Type: E102'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E103','E103','Int','Promoters - Cell Type: E103'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E104','E104','Int','Promoters - Cell Type: E104'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E105','E105','Int','Promoters - Cell Type: E105'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E106','E106','Int','Promoters - Cell Type: E106'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E107','E107','Int','Promoters - Cell Type: E107'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E108','E108','Int','Promoters - Cell Type: E108'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E109','E109','Int','Promoters - Cell Type: E109'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E110','E110','Int','Promoters - Cell Type: E110'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E111','E111','Int','Promoters - Cell Type: E111'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E112','E112','Int','Promoters - Cell Type: E112'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E113','E113','Int','Promoters - Cell Type: E113'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E114','E114','Int','Promoters - Cell Type: E114'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E115','E115','Int','Promoters - Cell Type: E115'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E116','E116','Int','Promoters - Cell Type: E116'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E117','E117','Int','Promoters - Cell Type: E117'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E118','E118','Int','Promoters - Cell Type: E118'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E119','E119','Int','Promoters - Cell Type: E119'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E120','E120','Int','Promoters - Cell Type: E120'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E121','E121','Int','Promoters - Cell Type: E121'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E122','E122','Int','Promoters - Cell Type: E122'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E123','E123','Int','Promoters - Cell Type: E123'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E124','E124','Int','Promoters - Cell Type: E124'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E125','E125','Int','Promoters - Cell Type: E125'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E126','E126','Int','Promoters - Cell Type: E126'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E127','E127','Int','Promoters - Cell Type: E127'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E128','E128','Int','Promoters - Cell Type: E128'),
 ('f_0002','va.chromHMM.promoters','va.chromHMM.promoters.E129','E129','Int','Promoters - Cell Type: E129'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E128','E128','Int','Enhancers - Cell Type: E128'),
 ('f_0001','va.chromHMM.enhancers','va.chromHMM.enhancers.E129','E129','Int','Enhancers - Cell Type: E129'),
 ('f_0004','va.dbNSFP','va.dbNSFP.aaref','aaref','String','reference amino acid; "." if the variant is a splicing site SNP (2bp on each end of an intron)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.aaalt','aaalt','String','alternative amino acid; "." if the variant is a splicing site SNP (2bp on each end of an intron)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.rs_dbsnp147','rs_dbsnp147','String','rs number from dbSNP 147'),
 ('f_0004','va.dbNSFP','va.dbNSFP.hg18_pos','hg18_pos','String','physical position on the chromosome as to hg18 (1-based coordinate)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.hg38_chr','hg38_chr','String','chromosome as to hg38, "." means the same as in the chr column'),
 ('f_0004','va.dbNSFP','va.dbNSFP.hg38_pos','hg38_pos','String','physical position on the chromosome as to hg38 (1-based coordinate)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.genename','genename','String','gene name; if the NScan be assigned to multiple genes, gene names are separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.uniprot_acc','uniprot_acc','String','Uniprot accession number. Multiple entries separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.uniprot_id','uniprot_id','String','Uniprot ID number. Multiple entries separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.uniprot_aapos','uniprot_aapos','String','amino acid position as to Uniprot. Multiple entries separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.interpro_domain','interpro_domain','String','domain or conserved site on which the variant locates. Domain annotations come from Interpro database. The number in the brackets following a specific domain is the count of times Interpro assigns the variant position to that domain, typically coming from different predicting databases. Multiple entries separated by ";".'),
 ('f_0004','va.dbNSFP','va.dbNSFP.cds_strand','cds_strand','String','coding sequence (CDS) strand (+ or -)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.refcodon','refcodon','String','reference codon'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SLR_test_statistic','SLR_test_statistic','String','SLR test statistic for testing natural selection on codons. A negative value indicates negative selection, and a positive value indicates positive selection. Larger magnitude of the value suggests stronger evidence.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.codonpos','codonpos','String','position on the codon (1, 2 or 3)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.fold_degenerate','fold_degenerate','String','degenerate type (0, 2 or 3)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.ancestral_allele','ancestral_allele','String','Ancestral allele (based on 1000 genomes reference data). The following comes from its original README file: ACTG - high-confidence call, ancestral state supproted by the other two sequences; actg - low-confindence call, ancestral state supported by one sequence only; N - failure, the ancestral state is not supported by any other sequence; - - the extant species contains an insertion at this postion; . - no coverage in the alignment'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Ensembl_geneid','Ensembl_geneid','String','Ensembl gene id'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Ensembl_transcriptid','Ensembl_transcriptid','String','Ensembl transcript ids (separated by ";")'),
 ('f_0004','va.dbNSFP','va.dbNSFP.aapos','aapos','String','amino acid position as to the protein "-1" if the variant is a splicing site SNP (2bp on each end of an intron)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.aapos_SIFT','aapos_SIFT','String','ENSP id and amino acid positions corresponding to SIFT scores. Multiple entries separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.aapos_FATHMM','aapos_FATHMM','String','ENSP id and amino acid positions corresponding to FATHMM scores. Multiple entries separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SIFT_score','SIFT_score','String','SIFT score (SIFTori). Scores range from 0 to 1. The smaller the score the more likely the SNP has damaging effect. Multiple scores separated by ";".'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SIFT_converted_rankscore','SIFT_converted_rankscore','String','SIFTori scores were first converted to SIFTnew=1-SIFTori, then ranked among all SIFTnew scores in dbNSFP. The rankscore is the ratio of the rank the SIFTnew score over the total number of SIFTnew scores in dbNSFP. If there are multiple scores,o nly the most damaging (largest) rankscore is presented. The rankscores range from 0.02654 to 0.87932.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SIFT_pred','SIFT_pred','String','If SIFTori is smaller than 0.05 (rankscore>0.55) the corresponding NS is predicted as "D(amaging)"; otherwise it is predicted as "T(olerated)". Multiple predictions separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Polyphen2_HDIV_score','Polyphen2_HDIV_score','String','Polyphen2 score based on HumDiv, i.e. hdiv_prob. The score ranges from 0 to 1. Multiple entries separated by ";".'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Polyphen2_HDIV_rankscore','Polyphen2_HDIV_rankscore','String','Polyphen2 HDIV scores were first ranked among all HDIV scores in dbNSFP. The rankscore is the ratio of the rank the score over the total number of the scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented. The scores range from 0.02656 to 0.89917.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Polyphen2_HVAR_score','Polyphen2_HVAR_score','String','Polyphen2 score based on HumVar, i.e. hvar_prob. The score ranges from 0 to 1. Multiple entries separated by \";\".'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Polyphen2_HVAR_rankscore','Polyphen2_HVAR_rankscore','String','Polyphen2 HVAR scores were first ranked among all HVAR scores in dbNSFP. The rankscore is the ratio of the rank the score over the total number of the scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented. The scores range from 0.01281 to 0.9711.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Polyphen2_HVAR_pred','Polyphen2_HVAR_pred','String','Polyphen2 prediction based on HumVar, "D" ("probably damaging", HVAR score in [0.909,1] or rankscore in [0.62955,0.9711]), "P" ("possibly damaging", HVAR in [0.447,0.908] or rankscore in [0.44359,0.62885]) and "B" ("benign", HVAR score in [0,0.446] or rankscore in [0.01281,0.44315]). Score cutoff for binary classification is 0.5 for HVAR score or 0.45998 for rankscore, i.e. the prediction is "neutral" if the HVAR score is smaller than 0.5 (rankscore is smaller than 0.45998), and "deleterious" if the HVAR score is larger than 0.5 (rankscore is larger than 0.45998). Multiple entries are separated by ";".'),
 ('f_0004','va.dbNSFP','va.dbNSFP.LRT_score','LRT_score','String','The original LRT two-sided p-value (LRTori), ranges from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.LRT_converted_rankscore','LRT_converted_rankscore','String','LRTori scores were first converted as LRTnew=1-LRTori*0.5 if Omega<1, or LRTnew=LRTori*0.5 if Omega>=1. Then LRTnew scores were ranked among all LRTnew scores in dbNSFP. The rankscore is the ratio of the rank over the total number of the scores in dbNSFP. The scores range from 0.00166 to 0.85682.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.LRT_pred','LRT_pred','String','LRT prediction, D(eleterious), N(eutral) or U(nknown), which is not solely determined by the score.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutationTaster_score','MutationTaster_score','String','MutationTaster p-value (MTori), ranges from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutationTaster_converted_rankscore','MutationTaster_converted_rankscore','String','The MTori scores were first converted: if the prediction is "A" or "D" MTnew=MTori; if the prediction is "N" or "P", MTnew=1-MTori. Then MTnew scores were ranked among all MTnew scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MTnew scores in dbNSFP. The scores range from 0.09067 to 0.80722.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutationTaster_pred','MutationTaster_pred','String','MutationTaster prediction, "A" ("disease_causing_automatic"), "D" ("disease_causing"), "N" ("polymorphism") or "P" ("polymorphism_automatic"). The score cutoff between "D" and "N" is 0.5 for MTori and 0.31655 for the rankscore.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutationAssessor_score','MutationAssessor_score','String','MutationAssessor functional impact combined score (MAori). The score ranges from -5.135 to 6.49 in dbNSFP. Please refer to Reva et al. (2011) Nucl. Acids Res. 39(17):e118 for details.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutationAssessor_rankscore','MutationAssessor_rankscore','String','MAori scores were ranked among all MAori scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MAori scores in dbNSFP. The scores range from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutationAssessor_pred','MutationAssessor_pred','String','MutationAssessor''s functional impact of a variant : predicted functional, i.e. high ("H") or medium ("M"), or predicted non-functional, i.e. low ("L") or neutral ("N"). The MAori score cutoffs between "H" and "M", "M" and "L", and "L" and "N", are 3.5, 1.935 and 0.8, respectively. The rankscore cutoffs between "H" and "M", "M" and "L", and "L" and "N", are 0.92924, 0.51945 and 0.19692, respectively.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.FATHMM_score','FATHMM_score','String','FATHMM default score (weighted for human inherited-disease mutations with Disease Ontology) (FATHMMori). Scores range from -18.09 to 11.0. The smaller the score the more likely the SNP has damaging effect. Multiple scores separated by ";" Please refer to Shihab et al. (2013) Human Mutation 34(1):57-65 for details.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.FATHMM_rankscore','FATHMM_rankscore','String','FATHMMnew=-FATHMMori scores were ranked among all FATHMMnew scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of FATHMMnew scores in dbNSFP. If there are multiple scores, only the most damaging (smallest) rankscore is presented. The scores range from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.FATHMM_pred','FATHMM_pred','String','If a FATHMMori score is <=-1.5 (or rankscore >=0.81415) the corresponding NS is predicted as "D(AMAGING)"; otherwise it is predicted as "T(OLERATED)". Multiple predictions separated by ";".'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MetaSVM_score','MetaSVM_score','String','Our support vector machine (SVM) based ensemble prediction score, which incorporated 10 scores (SIFT, PolyPhen-2 HDIV, PolyPhen-2 HVAR, GERP++, MutationTaster, Mutation Assessor, FATHMM, LRT, SiPhy, phyloP) and the maximum frequency observed in the 1000 genomes populations. Larger value means the SNV is more likely to be damaging. Scores range from -2 to 3 in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MetaSVM_rankscore','MetaSVM_rankscore','String','MetaSVM_rankscore: MetaSVM scores were ranked among all MetaSVM scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MetaSVM scores in dbNSFP. The scores range from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MetaSVM_pred','MetaSVM_pred','String','Prediction of our SVM based ensemble prediction score,"T(olerated)" or "D(amaging)". The score cutoff between "D" and "T" is 0. The rankscore cutoff between "D" and "T" is 0.83357.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MetaLR_score','MetaLR_score','String','Our logistic regression (LR) based ensemble prediction score, which incorporated 10 scores (SIFT, PolyPhen-2 HDIV, PolyPhen-2 HVAR, GERP++, MutationTaster, Mutation Assessor, FATHMM, LRT, SiPhy, phyloP) and the maximum frequency observed in the 1000 genomes populations. Larger value means the SNV is more likely to be damaging. Scores range from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MetaLR_rankscore','MetaLR_rankscore','String','MetaLR scores were ranked among all MetaLR scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MetaLR scores in dbNSFP. The scores range from 0 to 1.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MetaLR_pred','MetaLR_pred','String','Prediction of our MetaLR based ensemble prediction score,"T(olerated)" or "D(amaging)". The score cutoff between "D" and "T" is 0.5. The rankscore cutoff between "D" and "T" is 0.82268.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.reliability_index','reliability_index','String','Number of observed component scores (except the maximum frequency in the 1000 genomes populations) for MetaSVM and MetaLR. Ranges from 1 to 10. As MetaSVM and MetaLR scores are calculated based on imputed data, the less missing component scores, the higher the reliability of the scores and predictions.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.VEST3_score','VEST3_score','String','VEST 3.0 score. Score ranges from 0 to 1. The larger the score the more likely the mutation may cause functional change. In case there are multiple scores for the same variant, the largest score (most damaging) is presented. Please refer to Carter et al., (2013) BMC Genomics. 14(3) 1-16 for details. Please note this score is free for non-commercial use. For more details please refer to http://wiki.chasmsoftware.org/index.php/SoftwareLicense. Commercial users should contact the Johns Hopkins Technology Transfer office.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.VEST3_rankscore','VEST3_rankscore','String','VEST3 scores were ranked among all VEST3 scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of VEST3 scores in dbNSFP. The scores range from 0 to 1. Please note VEST score is free for non-commercial use. For more details please refer to http://wiki.chasmsoftware.org/index.php/SoftwareLicense. Commercial users should contact the Johns Hopkins Technology Transfer office.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.PROVEAN_score','PROVEAN_score','String','PROVEAN score (PROVEANori). Scores range from -14 to 14. The smaller the score the more likely the SNP has damaging effect. Multiple scores separated by ";". Details can be found in DOI: 10.1371/journal.pone.0046688'),
 ('f_0004','va.dbNSFP','va.dbNSFP.PROVEAN_converted_rankscore','PROVEAN_converted_rankscore','String','PROVEANori were first converted to PROVEANnew=1-(PROVEANori+14)/28, then ranked among all PROVEANnew scores in dbNSFP. The rankscore is the ratio of the rank the PROVEANnew score over the total number of PROVEANnew scores in dbNSFP. If there are multiple scores, only the most damaging (largest) rankscore is presented.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.PROVEAN_pred','PROVEAN_pred','String','If PROVEANori <= -2.5 (rankscore>=0.59) the corresponding NS is predicted as "D(amaging)"; otherwise it is predicted as "N(eutral)". Multiple predictions separated by ";"'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MCAP_score','MCAP_score','String','M-CAP score (details in DOI: 10.1038/ng.3703). Scores range from 0 to 1. The larger the score the more likely the SNP has damaging effect.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MCAP_rankscore','MCAP_rankscore','String','M-CAP scores were ranked among all M-CAP scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of M-CAP scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MCAP_pred','MCAP_pred','String','Prediction of M-CAP score based on the authors'' recommendation, "T(olerated)" or "D(amaging)". The score cutoff between "D" and "T" is 0.025.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.REVEL_score','REVEL_score','String','REVEL is an ensemble score based on 13 individual scores for predicting the pathogenicity of missense variants. Scores range from 0 to 1. The larger the score the more likely the SNP has damaging effect. "REVEL scores are freely available for non-commercial use. For other uses, please contact Weiva Sieh" (weiva.sieh@mssm.edu)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.REVEL_rankscore','REVEL_rankscore','String','REVEL scores were ranked among all REVEL scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of REVEL scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutPred_score','MutPred_score','String','General MutPred score. Scores range from 0 to 1. The larger the score the more likely the SNP has damaging effect.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutPred_rankscore','MutPred_rankscore','String','MutPred scores were ranked among all MutPred scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of MutPred scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutPred_protID','MutPred_protID','String','UniProt accession or Ensembl transcript ID used for MutPred_score calculation.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutPred_AAchange','MutPred_AAchange','String','Amino acid change used for MutPred_score calculation.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.MutPred_top5features','MutPred_top5features','String','Top 5 features (molecular mechanisms of disease) as predicted by MutPred with p values. MutPred_score > 0.5 and p < 0.05 are referred to as actionable hypotheses. MutPred_score > 0.75 and p < 0.05 are referred to as confident hypotheses. MutPred_score > 0.75 and p < 0.01 are referred to as very confident hypotheses.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.GERP_NR','GERP_NR','String','GERP++ neutral rate'),
 ('f_0004','va.dbNSFP','va.dbNSFP.GERP_RS','GERP_RS','String','GERP++ RS score, the larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.GERP_RS_rankscore','GERP_RS_rankscore','String','GERP++ RS scores were ranked among all GERP++ RS scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of GERP++ RS scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phyloP46way_primate','phyloP46way_primate','String','phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 10 primate genomes (including human). The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phyloP46way_primate_rankscore','phyloP46way_primate_rankscore','String','phyloP46way_primate scores were ranked among all phyloP46way_primate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP46way_primate scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phyloP46way_placental','phyloP46way_placental','String','phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 33 placental mammal genomes (including human). The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phyloP46way_placental_rankscore','phyloP46way_placental_rankscore','String','phyloP46way_placental scores were ranked among all phyloP46way_placental scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP46way_placental scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phyloP100way_vertebrate','phyloP100way_vertebrate','String','phyloP (phylogenetic p-values) conservation score based on the multiple alignments of 100 vertebrate genomes (including human). The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phyloP100way_vertebrate_rankscore','phyloP100way_vertebrate_rankscore','String','phyloP100way_vertebrate scores were ranked among all phyloP100way_vertebrate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phyloP100way_vertebrate scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phastCons46way_primate','phastCons46way_primate','String','phastCons conservation score based on the multiple alignments of 10 primate genomes (including human). The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phastCons46way_primate_rankscore','phastCons46way_primate_rankscore','String','phastCons46way_primate scores were ranked among all phastCons46way_primate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phastCons46way_primate scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phastCons46way_placental','phastCons46way_placental','String','phastCons conservation score based on the multiple alignments of 33 placental mammal genomes (including human). The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phastCons46way_placental_rankscore','phastCons46way_placental_rankscore','String','phastCons46way_placental scores were ranked among all phastCons46way_placental scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phastCons46way_placental scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phastCons100way_vertebrate','phastCons100way_vertebrate','String','phastCons conservation score based on the multiple alignments of 100 vertebrate genomes (including human). The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.phastCons100way_vertebrate_rankscore','phastCons100way_vertebrate_rankscore','String','phastCons100way_vertebrate scores were ranked among all phastCons100way_vertebrate scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of phastCons100way_vertebrate scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SiPhy_29way_pi','SiPhy_29way_pi','String','The estimated stationary distribution of A, C, G and T at the site, using SiPhy algorithm based on 29 mammals genomes.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SiPhy_29way_logOdds','SiPhy_29way_logOdds','String','SiPhy score based on 29 mammals genomes. The larger the score, the more conserved the site.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.SiPhy_29way_logOdds_rankscore','SiPhy_29way_logOdds_rankscore','String','SiPhy_29way_logOdds scores were ranked among all SiPhy_29way_logOdds scores in dbNSFP. The rankscore is the ratio of the rank of the score over the total number of SiPhy_29way_logOdds scores in dbNSFP.'),
 ('f_0004','va.dbNSFP','va.dbNSFP.LRT_omega','LRT_omega','String','estimated nonsynonymous-to-synonymous-rate ratio (Omega, reported by LRT)'),
 ('f_0004','va.dbNSFP','va.dbNSFP.clinvar_rs','clinvar_rs','String','rs number from the clinvar data set'),
 ('f_0004','va.dbNSFP','va.dbNSFP.clinvar_clinsig','clinvar_clinsig','String','clinical significance as to the clinvar data set 2 - Benign, 3 - Likely benign, 4 - Likely pathogenic, 5 - Pathogenic, 6 - drug response, 7 - histocompatibility. A negative score means the the score is for the ref allele'),
 ('f_0004','va.dbNSFP','va.dbNSFP.clinvar_trait','clinvar_trait','String','the trait/disease the clinvar_clnsig referring to'),
 ('f_0004','va.dbNSFP','va.dbNSFP.clinvar_golden_stars','clinvar_golden_stars','String','ClinVar Review Status summary: 0 - no assertion criteria provided, 1 - criteria provided, single submitter, 2 - criteria provided, multiple submitters, no conflicts, 3 - reviewed by expert panel, 4 - practice guideline'),
 ('f_0004','va.dbNSFP','va.dbNSFP.cosmic_id','cosmic_id','String','ID of the SNV at the COSMIC (Catalogue Of Somatic Mutations In Cancer) database'),
 ('f_0004','va.dbNSFP','va.dbNSFP.cosmic_count','cosmic_count','String','number of samples having this SNV in the COSMIC database'),
 ('f_0004','va.dbNSFP','va.dbNSFP.Polyphen2_HDIV_pred','Polyphen2_HDIV_pred','String','Polyphen2 prediction based on HumDiv, "D" ("porobably damaging", HDIV score in [0.957,1] or rankscore in [0.52996,0.89917]), "P" ("possibly damaging", HDIV score in [0.453,0.956] or rankscore in [0.34412,0.52842]) and "B" ("benign", HDIV score in [0,0.452] or rankscore in [0.02656,0.34399]). Score cutoff for binary classification is 0.5 for HDIV score or 0.35411 for rankscore, i.e. the prediction is "neutral" if the HDIV score is smaller than 0.5 (rankscore is smaller than 0.35411), and "deleterious" if the HDIV score is larger than 0.5 (rankscore is larger than 0.35411). Multiple entries are separated by ";".'),
 ('f_0005','va.dhs','va.dhs.TSS_Status','TSS_Status','Boolean','TSS (1=yes,0=no)'),
 ('f_0005','va.dhs','va.dhs.pvalue','pvalue','Double','P-value for replication'),
 ('f_0005','va.dhs','va.dhs.breast_vhmec','breast_vhmec','Boolean','Cell type: breast_vhmec'),
 ('f_0005','va.dhs','va.dhs.cd14_primary_cells','cd14_primary_cells','Boolean','Cell type: cd14_primary_cells'),
 ('f_0005','va.dhs','va.dhs.cd19_primary_cells','cd19_primary_cells','Boolean','Cell type: cd19_primary_cells'),
 ('f_0005','va.dhs','va.dhs.cd3_primary_cells','cd3_primary_cells','Boolean','Cell type: cd3_primary_cells'),
 ('f_0005','va.dhs','va.dhs.cd4_primary_cells','cd4_primary_cells','Boolean','Cell type: cd4_primary_cells'),
 ('f_0005','va.dhs','va.dhs.cd56_primary_cells','cd56_primary_cells','Boolean','Cell type: cd56_primary_cells'),
 ('f_0005','va.dhs','va.dhs.cd8_primary_cells','cd8_primary_cells','Boolean','Cell type: cd8_primary_cells'),
 ('f_0005','va.dhs','va.dhs.fetal_adrenal_gland','fetal_adrenal_gland','Boolean','Cell type: fetal_adrenal_gland'),
 ('f_0005','va.dhs','va.dhs.fetal_brain','fetal_brain','Boolean','Cell type: fetal_brain'),
 ('f_0005','va.dhs','va.dhs.fetal_heart','fetal_heart','Boolean','Cell type: fetal_heart'),
 ('f_0005','va.dhs','va.dhs.fetal_intestine_large','fetal_intestine_large','Boolean','Cell type: fetal_intestine_large'),
 ('f_0005','va.dhs','va.dhs.fetal_intestine_small','fetal_intestine_small','Boolean','Cell type: fetal_intestine_small'),
 ('f_0005','va.dhs','va.dhs.fetal_kidney','fetal_kidney','Boolean','Cell type: fetal_kidney'),
 ('f_0005','va.dhs','va.dhs.fetal_kidney_left','fetal_kidney_left','Boolean','Cell type: fetal_kidney_left'),
 ('f_0005','va.dhs','va.dhs.fetal_kidney_right','fetal_kidney_right','Boolean','Cell type: fetal_kidney_right'),
 ('f_0005','va.dhs','va.dhs.fetal_lung','fetal_lung','Boolean','Cell type: fetal_lung'),
 ('f_0005','va.dhs','va.dhs.fetal_lung_left','fetal_lung_left','Boolean','Cell type: fetal_lung_left'),
 ('f_0005','va.dhs','va.dhs.fetal_lung_right','fetal_lung_right','Boolean','Cell type: fetal_lung_right'),
 ('f_0005','va.dhs','va.dhs.fetal_muscle_arm','fetal_muscle_arm','Boolean','Cell type: fetal_muscle_arm'),
 ('f_0005','va.dhs','va.dhs.fetal_muscle_back','fetal_muscle_back','Boolean','Cell type: fetal_muscle_back'),
 ('f_0005','va.dhs','va.dhs.fetal_muscle_leg','fetal_muscle_leg','Boolean','Cell type: fetal_muscle_leg'),
 ('f_0005','va.dhs','va.dhs.fetal_muscle_trunk','fetal_muscle_trunk','Boolean','Cell type: fetal_muscle_trunk'),
 ('f_0005','va.dhs','va.dhs.fetal_placenta','fetal_placenta','Boolean','Cell type: fetal_placenta'),
 ('f_0005','va.dhs','va.dhs.fetal_renal_cortex','fetal_renal_cortex','Boolean','Cell type: fetal_renal_cortex'),
 ('f_0005','va.dhs','va.dhs.fetal_renal_cortex_left','fetal_renal_cortex_left','Boolean','Cell type: fetal_renal_cortex_left'),
 ('f_0005','va.dhs','va.dhs.fetal_renal_cortex_right','fetal_renal_cortex_right','Boolean','Cell type: fetal_renal_cortex_right'),
 ('f_0005','va.dhs','va.dhs.fetal_renal_pelvis','fetal_renal_pelvis','Boolean','Cell type: fetal_renal_pelvis'),
 ('f_0005','va.dhs','va.dhs.fetal_renal_pelvis_left','fetal_renal_pelvis_left','Boolean','Cell type: fetal_renal_pelvis_left'),
 ('f_0005','va.dhs','va.dhs.fetal_renal_pelvis_right','fetal_renal_pelvis_right','Boolean','Cell type: fetal_renal_pelvis_right'),
 ('f_0005','va.dhs','va.dhs.fetal_spinal_cord','fetal_spinal_cord','Boolean','Cell type: fetal_spinal_cord'),
 ('f_0005','va.dhs','va.dhs.fetal_stomach','fetal_stomach','Boolean','Cell type: fetal_stomach'),
 ('f_0005','va.dhs','va.dhs.fetal_testes','fetal_testes','Boolean','Cell type: fetal_testes'),
 ('f_0005','va.dhs','va.dhs.fetal_thymus','fetal_thymus','Boolean','Cell type: fetal_thymus'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_abdomen','fibroblasts_fetal_skin_abdomen','Boolean','Cell type: fibroblasts_fetal_skin_abdomen'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_back','fibroblasts_fetal_skin_back','Boolean','Cell type: fibroblasts_fetal_skin_back'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_biceps_left','fibroblasts_fetal_skin_biceps_left','Boolean','Cell type: fibroblasts_fetal_skin_biceps_left'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_biceps_right','fibroblasts_fetal_skin_biceps_right','Boolean','Cell type: fibroblasts_fetal_skin_biceps_right'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_quadriceps_left','fibroblasts_fetal_skin_quadriceps_left','Boolean','Cell type: fibroblasts_fetal_skin_quadriceps_left'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_quadriceps_right','fibroblasts_fetal_skin_quadriceps_right','Boolean','Cell type: fibroblasts_fetal_skin_quadriceps_right'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_scalp','fibroblasts_fetal_skin_scalp','Boolean','Cell type: fibroblasts_fetal_skin_scalp'),
 ('f_0005','va.dhs','va.dhs.fibroblasts_fetal_skin_upper_back','fibroblasts_fetal_skin_upper_back','Boolean','Cell type: fibroblasts_fetal_skin_upper_back'),
 ('f_0005','va.dhs','va.dhs.gastric','gastric','Boolean','Cell type: gastric'),
 ('f_0005','va.dhs','va.dhs.h1','h1','Boolean','Cell type: h1'),
 ('f_0005','va.dhs','va.dhs.h1_bmp4_derived_mesendoderm_cutured_cells','h1_bmp4_derived_mesendoderm_cultured_cells','Boolean','Cell type: h1_bmp4_derived_mesendoderm_cultured_cells'),
 ('f_0005','va.dhs','va.dhs.h1_bmp4_derived_trophoblast_cultured_cells','h1_bmp4_derived_trophoblast_cultured_cells','Boolean','Cell type: h1_bmp4_derived_trophoblast_cultured_cells'),
 ('f_0005','va.dhs','va.dhs.h1_derived_mesenchymal_stem_cells','h1_derived_mesenchymal_stem_cells','Boolean','Cell type: h1_derived_mesenchymal_stem_cells'),
 ('f_0005','va.dhs','va.dhs.h1_derived_neuronal_progenitor_cultured_cells','h1_derived_neuronal_progenitor_cultured_cells','Boolean','Cell type: h1_derived_neuronal_progenitor_cultured_cells'),
 ('f_0005','va.dhs','va.dhs.h9','h9','Boolean','Cell type: h9'),
 ('f_0005','va.dhs','va.dhs.imr90','imr90','Boolean','Cell type: imr90'),
 ('f_0005','va.dhs','va.dhs.mobilized_cd3_primary_cells','mobilized_cd3_primary_cells','Boolean','Cell type: mobilized_cd3_primary_cells'),
 ('f_0005','va.dhs','va.dhs.mobilized_cd34_primary_cells','mobilized_cd34_primary_cells','Boolean','Cell type: mobilized_cd34_primary_cells'),
 ('f_0005','va.dhs','va.dhs.mobilized_cd4_primary_cells','mobilized_cd4_primary_cells','Boolean','Cell type: mobilized_cd4_primary_cells'),
 ('f_0005','va.dhs','va.dhs.pancreas','pancreas','Boolean','Cell type: pancreas'),
 ('f_0005','va.dhs','va.dhs.penis_foreskin_fibroblast_primary_cells','penis_foreskin_fibroblast_primary_cells','Boolean','Cell type: penis_foreskin_fibroblast_primary_cells'),
 ('f_0005','va.dhs','va.dhs.penis_foreskin_keratinocyte_primary_cells','penis_foreskin_keratinocyte_primary_cells','Boolean','Cell type: penis_foreskin_keratinocyte_primary_cells'),
 ('f_0005','va.dhs','va.dhs.penis_foreskin_melanocyte_primary_cells','penis_foreskin_melanocyte_primary_cells','Boolean','Cell type: penis_foreskin_melanocyte_primary_cells'),
 ('f_0006','va.discovEHR','va.discovEHR.AF','AF','Double','Allele Frequency in discovEHR. AF=0.001 is equal to AF<=0.001'),
 ('f_0007','va.eigen','va.eigen.raw','raw','Double','Eigen score raw'),
 ('f_0007','va.eigen','va.eigen.phred','phred','Double','Eigen score phred transformed'),
 ('f_0007','va.eigen','va.eigen.PC_raw','PC_raw','Double','Eigen PC score raw'),
 ('f_0007','va.eigen','va.eigen.PC_phred','PC_phred','Double','Eigen PC score phred transformed'),
 ('f_0008','va.fantom5','va.fantom5.robust','robust','Boolean','FANTOM5 enhancers - robust'),
 ('f_0021','va.fantom5','va.fantom5.permissive','permissive','Boolean','FANTOM5 enhancers - permissive'),
 ('f_0009','va.funseq2','va.funseq2.score','score','Double','FUNSEQ2 Score'),
 ('f_0010','va.gencode19','va.gencode19.three_prime_utr','three_prime_utr','Boolean','3'' UTR Exons'),
 ('f_0010','va.gencode19','va.gencode19.five_prime_utr','five_prime_utr','Boolean','5'' UTR Exons'),
 ('f_0010','va.gencode19','va.gencode19.exonsc','exonsc','Boolean','Coding exons'),
 ('f_0010','va.gencode19','va.gencode19.coding15','coding15','Boolean','All exons (including untranslated sequences) plus 15bp at each end'),
 ('f_0010','va.gencode19','va.gencode19.introns','introns','Boolean','Introns'),
 ('f_0010','va.gencode19','va.gencode19.exonsa','exonsa','Boolean','All exons (including untranslated sequences)'),
 ('f_0010','va.gencode19','va.gencode19.promotersreg','promotersreg','Boolean','200 bp upstream of the TSS, promoters-flanking regions'),
 ('f_0011','va.gene.names','va.gene.names.gene_symbol','gene_symbol','String','Gene symbol from HGNC'),
 ('f_0011','va.gene.names','va.gene.names.ensembl_gene','ensembl_gene','String','Ensembl gene id (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.chromosome','chromosome','String','Chromosome number (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.gene_old_names','gene_old_names','String','Old gene sybmol (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.gene_other_names','gene_other_names','String','Other gene names (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.uniprot_acc','uniprot_acc','String','Uniprot acc number (from HGNC and Uniprot)'),
 ('f_0011','va.gene.names','va.gene.names.uniprot_ID','uniprot_ID','String','Uniprot id (from HGNC and Uniprot)'),
 ('f_0011','va.gene.names','va.gene.names.entrez_gene_ID','entrez_gene_ID','String','Entrez gene id (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.ccds_ID','ccds_ID','String','CCDS id (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.refseq_ID','refseq_ID','String','Refseq gene id (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.ucsc_ID','ucsc_ID','String','UCSC gene id (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.mim_ID','mim_ID','String','MIM gene id (from HGNC)'),
 ('f_0011','va.gene.names','va.gene.names.gene_full_name','gene_full_name','String','Gene full name (from HGNC)'),
 ('f_0011','va.gene.constraint','va.gene.constraint.gene_symbol','gene_symbol','String','Ensembl gene symbol'),
 ('f_0011','va.gene.constraint','va.gene.constraint.transcript_ID','transcript_ID','String','Ensembl transcript ID'),
 ('f_0011','va.gene.constraint','va.gene.constraint.chromosome','chromosome','String','Chromosome'),
 ('f_0011','va.gene.constraint','va.gene.constraint.n_exons','n_exons','Int','Number of exons in the transcript'),
 ('f_0011','va.gene.constraint','va.gene.constraint.n_bp','n_bp','Int','Number of coding base pairs'),
 ('f_0011','va.gene.constraint','va.gene.constraint.syn_z','syn_z','Double','Corrected synonymous Z score'),
 ('f_0011','va.gene.constraint','va.gene.constraint.mis_z','mis_z','Double','Corrected missense Z score'),
 ('f_0011','va.gene.constraint','va.gene.constraint.lof_z','lof_z','Double','Corrected loss-of-function Z score'),
 ('f_0011','va.gene.constraint','va.gene.constraint.pli','pli','Double','The probability of being loss-of-function intolerant (intolerant of both heterozygous and homozygous lof variants)'),
 ('f_0011','va.gene.constraint','va.gene.constraint.syn_z_notcga','syn_z_notcga','Double','Corrected synonymous Z score: no TCGA'),
 ('f_0011','va.gene.constraint','va.gene.constraint.mis_z_notcga','mis_z_notcga','Double','Corrected missense Z score: no TCGA'),
 ('f_0011','va.gene.constraint','va.gene.constraint.lof_z_notcga','lof_z_notcga','Double','Corrected loss-of-function Z score: no TCGA'),
 ('f_0011','va.gene.constraint','va.gene.constraint.pli_notcga','pli_notcga','Double','The probability of being loss-of-function intolerant (intolerant of both heterozygous and homozygous lof variants): no TCGA'),
 ('f_0011','va.gene.constraint','va.gene.constraint.syn_z_nopsych','syn_z_nopsych','Double','Corrected synonymous Z score: no psych'),
 ('f_0011','va.gene.constraint','va.gene.constraint.mis_z_nopsych','mis_z_nopsych','Double','Corrected missense Z score: no psych'),
 ('f_0011','va.gene.constraint','va.gene.constraint.lof_z_nopsych','lof_z_nopsych','Double','Corrected loss-of-function Z score: no psych'),
 ('f_0011','va.gene.constraint','va.gene.constraint.pli_nopsych','pli_nopsych','Double','The probability of being loss-of-function intolerant (intolerant of both heterozygous and homozygous lof variants): no psych'),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.rsid','rsid','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.qual','qual','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.filters','filters','Set[String]',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.pass','pass','Boolean',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC','AC','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF','AF','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.BaseQRankSum','BaseQRankSum','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.ClippingRankSum','ClippingRankSum','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.DB','DB','Boolean',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.DP','DP','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.FS','FS','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.InbreedingCoeff','InbreedingCoeff','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.MQ','MQ','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.MQRankSum','MQRankSum','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.QD','QD','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.ReadPosRankSum','ReadPosRankSum','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.SOR','SOR','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.VQSLOD','VQSLOD','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.VQSR_culprit','VQSR_culprit','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.VQSR_NEGATIVE_TRAIN_SITE','VQSR_NEGATIVE_TRAIN_SITE','Boolean',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.VQSR_POSITIVE_TRAIN_SITE','VQSR_POSITIVE_TRAIN_SITE','Boolean',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GQ_HIST_ALT','GQ_HIST_ALT','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.DP_HIST_ALT','DP_HIST_ALT','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AB_HIST_ALT','AB_HIST_ALT','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GQ_HIST_ALL','GQ_HIST_ALL','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.DP_HIST_ALL','DP_HIST_ALL','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AB_HIST_ALL','AB_HIST_ALL','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_AFR','AC_AFR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_AMR','AC_AMR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_ASJ','AC_ASJ','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_EAS','AC_EAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_FIN','AC_FIN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_NFE','AC_NFE','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_OTH','AC_OTH','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_SAS','AC_SAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_Male','AC_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_Female','AC_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_AFR','AN_AFR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_AMR','AN_AMR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_ASJ','AN_ASJ','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_EAS','AN_EAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_FIN','AN_FIN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_NFE','AN_NFE','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_OTH','AN_OTH','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_SAS','AN_SAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_Male','AN_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_Female','AN_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_AFR','AF_AFR','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_AMR','AF_AMR','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_ASJ','AF_ASJ','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_EAS','AF_EAS','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_FIN','AF_FIN','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_NFE','AF_NFE','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_OTH','AF_OTH','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_SAS','AF_SAS','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_Male','AF_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_Female','AF_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_AFR','GC_AFR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_AMR','GC_AMR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_ASJ','GC_ASJ','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_EAS','GC_EAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_FIN','GC_FIN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_NFE','GC_NFE','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_OTH','GC_OTH','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_SAS','GC_SAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_Male','GC_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_Female','GC_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_raw','AC_raw','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_raw','AN_raw','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_raw','AF_raw','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_raw','GC_raw','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC','GC','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_AFR','Hom_AFR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_AMR','Hom_AMR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_ASJ','Hom_ASJ','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_EAS','Hom_EAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_FIN','Hom_FIN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_NFE','Hom_NFE','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_OTH','Hom_OTH','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_SAS','Hom_SAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_Male','Hom_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_Female','Hom_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom_raw','Hom_raw','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hom','Hom','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.STAR_AC','STAR_AC','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.STAR_AC_raw','STAR_AC_raw','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.STAR_Hom','STAR_Hom','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.POPMAX','POPMAX','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_POPMAX','AC_POPMAX','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_POPMAX','AN_POPMAX','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_POPMAX','AF_POPMAX','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.DP_MEDIAN','DP_MEDIAN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.DREF_MEDIAN','DREF_MEDIAN','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GQ_MEDIAN','GQ_MEDIAN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AB_MEDIAN','AB_MEDIAN','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AS_RF','AS_RF','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AS_FilterStatus','AS_FilterStatus','Set[String]',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AS_RF_POSITIVE_TRAIN','AS_RF_POSITIVE_TRAIN','Array[Int]',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AS_RF_NEGATIVE_TRAIN','AS_RF_NEGATIVE_TRAIN','Array[Int]',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.CSQ','CSQ','Array[String]',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_FIN_Male','AN_FIN_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_EAS_Female','AN_EAS_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_NFE_Female','AN_NFE_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_AFR_Male','AC_AFR_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_AMR_Female','AN_AMR_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_AMR_Male','AF_AMR_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_NFE','Hemi_NFE','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_AFR','Hemi_AFR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_ASJ_Female','AC_ASJ_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_FIN_Female','AF_FIN_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_ASJ_Male','AN_ASJ_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_OTH_Female','AC_OTH_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_OTH_Male','GC_OTH_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_FIN_Male','GC_FIN_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_NFE_Female','AC_NFE_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_EAS_Male','AC_EAS_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_OTH_Male','AC_OTH_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_SAS_Male','GC_SAS_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_AMR','Hemi_AMR','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_NFE_Male','AC_NFE_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi','Hemi','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_FIN_Female','AN_FIN_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_EAS_Male','GC_EAS_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_ASJ_Female','GC_ASJ_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_SAS_Female','GC_SAS_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_ASJ_Male','GC_ASJ_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_SAS','Hemi_SAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_ASJ_Female','AN_ASJ_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_FIN_Male','AF_FIN_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_OTH_Male','AN_OTH_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_AFR_Male','AF_AFR_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.STAR_Hemi','STAR_Hemi','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_SAS_Male','AF_SAS_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_ASJ','Hemi_ASJ','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_SAS_Female','AN_SAS_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_AFR_Female','AN_AFR_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_raw','Hemi_raw','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_OTH_Male','AF_OTH_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_SAS_Female','AC_SAS_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_NFE_Female','AF_NFE_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_EAS_Female','AF_EAS_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_OTH_Female','AN_OTH_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_EAS_Male','AF_EAS_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_SAS_Female','AF_SAS_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_AFR_Female','GC_AFR_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_AFR_Female','AF_AFR_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_FIN_Female','AC_FIN_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_OTH','Hemi_OTH','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_AMR_Male','GC_AMR_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_AFR_Female','AC_AFR_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_NFE_Male','GC_NFE_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_AMR_Female','AF_AMR_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_NFE_Female','GC_NFE_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_AFR_Male','AN_AFR_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_NFE_Male','AN_NFE_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_AMR_Male','AC_AMR_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_AMR_Female','GC_AMR_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_SAS_Male','AC_SAS_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_ASJ_Male','AF_ASJ_Male','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_FIN_Female','GC_FIN_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_EAS_Female','AC_EAS_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_AMR_Female','AC_AMR_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_FIN','Hemi_FIN','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_FIN_Male','AC_FIN_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_EAS_Female','GC_EAS_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_ASJ_Female','AF_ASJ_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AF_OTH_Female','AF_OTH_Female','Double',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_AFR_Male','GC_AFR_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_SAS_Male','AN_SAS_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.GC_OTH_Female','GC_OTH_Female','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_AMR_Male','AN_AMR_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.Hemi_EAS','Hemi_EAS','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AC_ASJ_Male','AC_ASJ_Male','Int',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN_EAS_Male','AN_EAS_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.rsid','rsid','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.qual','qual','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.filters','filters','Set[String]',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.pass','pass','Boolean',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC','AC','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF','AF','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.BaseQRankSum','BaseQRankSum','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.ClippingRankSum','ClippingRankSum','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.DB','DB','Boolean',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.DP','DP','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.FS','FS','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.InbreedingCoeff','InbreedingCoeff','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.MQ','MQ','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.MQRankSum','MQRankSum','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.QD','QD','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.ReadPosRankSum','ReadPosRankSum','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.SOR','SOR','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.VQSLOD','VQSLOD','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.VQSR_culprit','VQSR_culprit','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.VQSR_NEGATIVE_TRAIN_SITE','VQSR_NEGATIVE_TRAIN_SITE','Boolean',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.VQSR_POSITIVE_TRAIN_SITE','VQSR_POSITIVE_TRAIN_SITE','Boolean',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GQ_HIST_ALT','GQ_HIST_ALT','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.DP_HIST_ALT','DP_HIST_ALT','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AB_HIST_ALT','AB_HIST_ALT','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GQ_HIST_ALL','GQ_HIST_ALL','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.DP_HIST_ALL','DP_HIST_ALL','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AB_HIST_ALL','AB_HIST_ALL','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_AFR','AC_AFR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_AMR','AC_AMR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_ASJ','AC_ASJ','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_EAS','AC_EAS','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_FIN','AC_FIN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_NFE','AC_NFE','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_OTH','AC_OTH','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_Male','AC_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_Female','AC_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_AFR','AN_AFR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_AMR','AN_AMR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_ASJ','AN_ASJ','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_EAS','AN_EAS','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_FIN','AN_FIN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_NFE','AN_NFE','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_OTH','AN_OTH','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_Male','AN_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_Female','AN_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_AFR','AF_AFR','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_AMR','AF_AMR','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_ASJ','AF_ASJ','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_EAS','AF_EAS','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_FIN','AF_FIN','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_NFE','AF_NFE','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_OTH','AF_OTH','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_Male','AF_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_Female','AF_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_AFR','GC_AFR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_AMR','GC_AMR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_ASJ','GC_ASJ','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_EAS','GC_EAS','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_FIN','GC_FIN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_NFE','GC_NFE','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_OTH','GC_OTH','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_Male','GC_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_Female','GC_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_raw','AC_raw','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_raw','AN_raw','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_raw','AF_raw','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_raw','GC_raw','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC','GC','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_AFR','Hom_AFR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_AMR','Hom_AMR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_ASJ','Hom_ASJ','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_EAS','Hom_EAS','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_FIN','Hom_FIN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_NFE','Hom_NFE','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_OTH','Hom_OTH','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_Male','Hom_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_Female','Hom_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom_raw','Hom_raw','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hom','Hom','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.STAR_AC','STAR_AC','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.STAR_AC_raw','STAR_AC_raw','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.STAR_Hom','STAR_Hom','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.POPMAX','POPMAX','String',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_POPMAX','AC_POPMAX','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_POPMAX','AN_POPMAX','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_POPMAX','AF_POPMAX','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.DP_MEDIAN','DP_MEDIAN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.DREF_MEDIAN','DREF_MEDIAN','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GQ_MEDIAN','GQ_MEDIAN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AB_MEDIAN','AB_MEDIAN','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AS_RF','AS_RF','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AS_FilterStatus','AS_FilterStatus','Set[String]',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AS_RF_POSITIVE_TRAIN','AS_RF_POSITIVE_TRAIN','Array[Int]',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AS_RF_NEGATIVE_TRAIN','AS_RF_NEGATIVE_TRAIN','Array[Int]',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.CSQ','CSQ','Array[String]',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_FIN_Male','AN_FIN_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_EAS_Female','AN_EAS_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_NFE_Female','AN_NFE_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_AFR_Male','AC_AFR_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_AMR_Female','AN_AMR_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_AMR_Male','AF_AMR_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_NFE','Hemi_NFE','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_AFR','Hemi_AFR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_ASJ_Female','AC_ASJ_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_FIN_Female','AF_FIN_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_ASJ_Male','AN_ASJ_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_OTH_Female','AC_OTH_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_OTH_Male','GC_OTH_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_FIN_Male','GC_FIN_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_NFE_Female','AC_NFE_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_EAS_Male','AC_EAS_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_OTH_Male','AC_OTH_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_AMR','Hemi_AMR','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_NFE_Male','AC_NFE_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi','Hemi','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_FIN_Female','AN_FIN_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_EAS_Male','GC_EAS_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_ASJ_Female','GC_ASJ_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_ASJ_Male','GC_ASJ_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_ASJ_Female','AN_ASJ_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_FIN_Male','AF_FIN_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_OTH_Male','AN_OTH_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_AFR_Male','AF_AFR_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.STAR_Hemi','STAR_Hemi','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_ASJ','Hemi_ASJ','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_AFR_Female','AN_AFR_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_raw','Hemi_raw','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_OTH_Male','AF_OTH_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_NFE_Female','AF_NFE_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_EAS_Female','AF_EAS_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_OTH_Female','AN_OTH_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_EAS_Male','AF_EAS_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_AFR_Female','GC_AFR_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_AFR_Female','AF_AFR_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_FIN_Female','AC_FIN_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_OTH','Hemi_OTH','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_AMR_Male','GC_AMR_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_AFR_Female','AC_AFR_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_NFE_Male','GC_NFE_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_AMR_Female','AF_AMR_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_NFE_Female','GC_NFE_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_AFR_Male','AN_AFR_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_NFE_Male','AN_NFE_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_AMR_Male','AC_AMR_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_AMR_Female','GC_AMR_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_ASJ_Male','AF_ASJ_Male','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_FIN_Female','GC_FIN_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_EAS_Female','AC_EAS_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_AMR_Female','AC_AMR_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_FIN','Hemi_FIN','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_FIN_Male','AC_FIN_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_EAS_Female','GC_EAS_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_ASJ_Female','AF_ASJ_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AF_OTH_Female','AF_OTH_Female','Double',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_AFR_Male','GC_AFR_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.GC_OTH_Female','GC_OTH_Female','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_AMR_Male','AN_AMR_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.Hemi_EAS','Hemi_EAS','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AC_ASJ_Male','AC_ASJ_Male','Int',''),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN_EAS_Male','AN_EAS_Male','Int',''),
 ('f_0015','va','va.isLCR','isLCR','Boolean',''),
 ('f_0016','va.ldscore','va.ldscore.coding_ucsc','coding_ucsc','Boolean','Annotations from RefSeq gene models were obtained from the UCSC Genome Browser and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.conserved','conserved','Boolean','Regions conserved in mammals were obtained from Lindblad-Toh et al  and post-processed by Ward and Kelli.'),
 ('f_0016','va.ldscore','va.ldscore.ctcf','ctcf','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the union over the six cell lines.'),
 ('f_0016','va.ldscore','va.ldscore.dgf','dgf','Boolean','DNaseI Digital Genomic Footprinting from ENCODE3 and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.dhsp','dhsp','Boolean','DHSs are a combination of ENCODE and Roadmap Epigenomics data, post-processed by Trynka et al. Union of cell type–specific annotations.'),
 ('f_0016','va.ldscore','va.ldscore.dhs','dhs','Boolean','DHSs are a combination of ENCODE and Roadmap Epigenomics data, post-processed by Trynka et al. Union of cell type–specific annotations.'),
 ('f_0016','va.ldscore','va.ldscore.enhancers','enhancers','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the union over the six cell lines.'),
 ('f_0016','va.ldscore','va.ldscore.fantom5','fantom5','Boolean','FANTOM5 enhancers were obtained from Andersson et al.'),
 ('f_0016','va.ldscore','va.ldscore.dhsf','dhsf','Boolean','DHSs are a combination of ENCODE and Roadmap Epigenomics data, post-processed by Trynka et al. Union of cell type–specific annotations.'),
 ('f_0016','va.ldscore','va.ldscore.h3k4me1p','h3k4me1p','Boolean','Cell type–specific H3K4me1, H3K4me3 and H3K9ac data were all obtained from Roadmap Epigenomics and post-processed by Trynka et al. For each mark, we took a union over cell types'),
 ('f_0016','va.ldscore','va.ldscore.h3k4me1','h3k4me1','Boolean','Cell type–specific H3K4me1, H3K4me3 and H3K9ac data were all obtained from Roadmap Epigenomics and post-processed by Trynka et al. For each mark, we took a union over cell types'),
 ('f_0016','va.ldscore','va.ldscore.h3k4me3p','h3k4me3p','Boolean','Cell type–specific H3K4me1, H3K4me3 and H3K9ac data were all obtained from Roadmap Epigenomics and post-processed by Trynka et al. For each mark, we took a union over cell types'),
 ('f_0016','va.ldscore','va.ldscore.h3k4me3','h3k4me3','Boolean','Cell type–specific H3K4me1, H3K4me3 and H3K9ac data were all obtained from Roadmap Epigenomics and post-processed by Trynka et al. For each mark, we took a union over cell types'),
 ('f_0016','va.ldscore','va.ldscore.h3k9acp','h3k9acp','Boolean','Cell type–specific H3K4me1, H3K4me3 and H3K9ac data were all obtained from Roadmap Epigenomics and post-processed by Trynka et al. For each mark, we took a union over cell types'),
 ('f_0016','va.ldscore','va.ldscore.h3k9ac','h3k9ac','Boolean','Cell type–specific H3K4me1, H3K4me3 and H3K9ac data were all obtained from Roadmap Epigenomics and post-processed by Trynka et al. For each mark, we took a union over cell types'),
 ('f_0016','va.ldscore','va.ldscore.h3k27ac','h3k27ac','Boolean','H3K27ac data was obtained from Hnisz et al.'),
 ('f_0016','va.ldscore','va.ldscore.intron_ucsc','intron_ucsc','Boolean','Annotations from RefSeq gene models were obtained from the UCSC Genome Browser and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.promoter_ucsc','promoter_ucsc','Boolean','Annotations from RefSeq gene models were obtained from the UCSC Genome Browser and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.promoter_flanking','promoter_flanking','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the union over the six cell lines.'),
 ('f_0016','va.ldscore','va.ldscore.repressed','repressed','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the intersection over the six cell lines.'),
 ('f_0016','va.ldscore','va.ldscore.superenh','superenh','Boolean','Super Enhancers data was obtained from Hnisz et al.'),
 ('f_0016','va.ldscore','va.ldscore.tfbs','tfbs','Boolean','Transcription factor binding sites were obtained from ENCODE3 and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.transcribed','transcribed','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the union over the six cell lines.'),
 ('f_0016','va.ldscore','va.ldscore.tss','tss','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the union over the six cell lines.'),
 ('f_0016','va.ldscore','va.ldscore.three_prime_utr','three_prime_utr','Boolean','Annotations from RefSeq gene models were obtained from the UCSC Genome Browser and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.five_prime_utr','five_prime_utr','Boolean','Annotations from RefSeq gene models were obtained from the UCSC Genome Browser and post-processed by Gusev et al.'),
 ('f_0016','va.ldscore','va.ldscore.weakenh','weakenh','Boolean','The combined chromHMM and Segway annotations for six cell lines were obtained from Hoffman et al. This annotation is the union over the six cell lines.'),
 ('f_0017','va.linsight','va.linsight.score','score','Double','The probability of fitness consequences for mutations at each site i'),
 ('f_0018','va.segway','va.segway.sum_score','sum_score','Double','Sum of functionality score'),
 ('f_0018','va.segway','va.segway.mean_score','mean_score','Double','Average base-wise functionality score'),
 ('f_0018','va.segway','va.segway.a549','a549','String','Cell type: a549'),
 ('f_0018','va.segway','va.segway.adipose_nuclei','adipose_nuclei','String','Cell type: adipose_nuclei'),
 ('f_0018','va.segway','va.segway.adult_liver','adult_liver','String','Cell type: adult_liver'),
 ('f_0018','va.segway','va.segway.ag04449','ag04449','String','Cell type: ag04449'),
 ('f_0018','va.segway','va.segway.ag04450','ag04450','String','Cell type: ag04450'),
 ('f_0018','va.segway','va.segway.ag09309','ag09309','String','Cell type: ag09309'),
 ('f_0018','va.segway','va.segway.ag09319','ag09319','String','Cell type: ag09319'),
 ('f_0018','va.segway','va.segway.ag10803','ag10803','String','Cell type: ag10803'),
 ('f_0018','va.segway','va.segway.aoaf','aoaf','String','Cell type: aoaf'),
 ('f_0018','va.segway','va.segway.aorta','aorta','String','Cell type: aorta'),
 ('f_0018','va.segway','va.segway.bc_colon_h12817n','bc_colon_h12817n','String','Cell type: bc_colon_h12817n'),
 ('f_0018','va.segway','va.segway.bc_lung_01_11002','bc_lung_01_11002','String','Cell type: bc_lung_01_11002'),
 ('f_0020','va.vista','va.vista.location','location','String','VISTA Enhancers: Annotated expression patterns'),
 ('f_0019','va.tfbs','va.tfbs.gene','gene','String','Transcription factor binding sites from ENCODE v3'),
 ('f_0018','va.segway','va.segway.bc_lung_h12817n','bc_lung_h12817n','String','Cell type: bc_lung_h12817n'),
 ('f_0018','va.segway','va.segway.bc_spleen_h12817n','bc_spleen_h12817n','String','Cell type: bc_spleen_h12817n'),
 ('f_0018','va.segway','va.segway.be2_c','be2_c','String','Cell type: be2_c'),
 ('f_0018','va.segway','va.segway.bj','bj','String','Cell type: bj'),
 ('f_0018','va.segway','va.segway.bone_marrow_derived_mesenchymal_stem_cell_cultured_cells','bone_marrow_derived_mesenchymal_stem_cell_cultured_cells','String','Cell type: bone_marrow_derived_mesenchymal_stem_cell_cultured_cells'),
 ('f_0018','va.segway','va.segway.brain_angular_gyrus','brain_angular_gyrus','String','Cell type: brain_angular_gyrus'),
 ('f_0018','va.segway','va.segway.brain_anterior_caudate','brain_anterior_caudate','String','Cell type: brain_anterior_caudate'),
 ('f_0018','va.segway','va.segway.brain_cingulate_gyrus','brain_cingulate_gyrus','String','Cell type: brain_cingulate_gyrus'),
 ('f_0018','va.segway','va.segway.brain_hippocampus_middle','brain_hippocampus_middle','String','Cell type: brain_hippocampus_middle'),
 ('f_0018','va.segway','va.segway.brain_inferior_temporal_lobe','brain_inferior_temporal_lobe','String','Cell type: brain_inferior_temporal_lobe'),
 ('f_0018','va.segway','va.segway.brain_mid_frontal_lobe','brain_mid_frontal_lobe','String','Cell type: brain_mid_frontal_lobe'),
 ('f_0018','va.segway','va.segway.brain_substantia_nigra','brain_substantia_nigra','String','Cell type: brain_substantia_nigra'),
 ('f_0018','va.segway','va.segway.breast_myoepithelial_cells','breast_myoepithelial_cells','String','Cell type: breast_myoepithelial_cells'),
 ('f_0018','va.segway','va.segway.breast_vhmec','breast_vhmec','String','Cell type: breast_vhmec'),
 ('f_0018','va.segway','va.segway.caco_2','caco_2','String','Cell type: caco_2'),
 ('f_0018','va.segway','va.segway.cd14_primary_cells','cd14_primary_cells','String','Cell type: cd14_primary_cells'),
 ('f_0018','va.segway','va.segway.cd19_primary_cells_peripheral_uw','cd19_primary_cells_peripheral_uw','String','Cell type: cd19_primary_cells_peripheral_uw'),
 ('f_0018','va.segway','va.segway.cd3_primary_cells_cord_bi','cd3_primary_cells_cord_bi','String','Cell type: cd3_primary_cells_cord_bi'),
 ('f_0018','va.segway','va.segway.cd3_primary_cells_peripheral_uw','cd3_primary_cells_peripheral_uw','String','Cell type: cd3_primary_cells_peripheral_uw'),
 ('f_0018','va.segway','va.segway.cd4_cd25_cd127_treg_primary_cells','cd4_cd25_cd127_treg_primary_cells','String','Cell type: cd4+_cd25+_cd127-_treg_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_cd25_cd45ra_naive_primary_cells','cd4_cd25_cd45ra_naive_primary_cells','String','Cell type: cd4+_cd25-_cd45ra+_naive_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_cd25_cd45ro_memory_primary_cells','cd4_cd25_cd45ro_memory_primary_cells','String','Cell type: cd4_cd25_cd45ro_memory_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_cd25_il17_pma_ionomcyin_stimulated_th17_primary_cells','cd4_cd25_il17_pma_ionomcyin_stimulated_th17_primary_cells','String','Cell type: cd4+_cd25-_il17+_pma_ionomcyin_stimulated_th17_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_cd25_il17_pma_ionomycin_stimulated_macs_purified_th_primary_cells','cd4_cd25_il17_pma_ionomycin_stimulated_macs_purified_th_primary_cells','String','Cell type: cd4+_cd25-_il17-_pma_ionomycin_stimulated_macs_purified_th_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_cd25int_cd127_tmem_primary_cells','cd4_cd25int_cd127_tmem_primary_cells','String','Cell type: cd4+_cd25int_cd127+_tmem_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_cd25_th_primary_cells','cd4_cd25_th_primary_cells','String','Cell type: cd4+_cd25-_th_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_memory_primary_cells','cd4_memory_primary_cells','String','Cell type: cd4_memory_primary_cells'),
 ('f_0018','va.segway','va.segway.cd4_naive_primary_cells','cd4_naive_primary_cells','String','Cell type: cd4_naive_primary_cells'),
 ('f_0018','va.segway','va.segway.cd56_primary_cells','cd56_primary_cells','String','Cell type: cd56_primary_cells'),
 ('f_0018','va.segway','va.segway.cd8_memory_primary_cells','cd8_memory_primary_cells','String','Cell type: cd8_memory_primary_cells'),
 ('f_0018','va.segway','va.segway.cd8_naive_primary_cells','cd8_naive_primary_cells','String','Cell type: cd8_naive_primary_cells'),
 ('f_0018','va.segway','va.segway.chondrocytes_from_bone_marrow_derived_mesenchymal_stem_cell_cultured_cells','chondrocytes_from_bone_marrow_derived_mesenchymal_stem_cell_cultured_cells','String','Cell type: chondrocytes_from_bone_marrow_derived_mesenchymal_stem_cell_cultured_cells'),
 ('f_0018','va.segway','va.segway.colonic_mucosa','colonic_mucosa','String','Cell type: colonic_mucosa'),
 ('f_0018','va.segway','va.segway.colon_smooth_muscle','colon_smooth_muscle','String','Cell type: colon_smooth_muscle'),
 ('f_0018','va.segway','va.segway.dnd41','dnd41','String','Cell type: dnd41'),
 ('f_0018','va.segway','va.segway.duodenum_mucosa','duodenum_mucosa','String','Cell type: duodenum_mucosa'),
 ('f_0018','va.segway','va.segway.duodenum_smooth_muscle','duodenum_smooth_muscle','String','Cell type: duodenum_smooth_muscle'),
 ('f_0018','va.segway','va.segway.ecc_1','ecc_1','String','Cell type: ecc_1'),
 ('f_0018','va.segway','va.segway.es_i3_cell_line','es_i3_cell_line','String','Cell type: es_i3_cell_line'),
 ('f_0018','va.segway','va.segway.es_wa7_cell_line','es_wa7_cell_line','String','Cell type: es_wa7_cell_line'),
 ('f_0018','va.segway','va.segway.fetal_adrenal_gland','fetal_adrenal_gland','String','Cell type: fetal_adrenal_gland'),
 ('f_0018','va.segway','va.segway.fetal_brain_female','fetal_brain_female','String','Cell type: fetal_brain_female'),
 ('f_0018','va.segway','va.segway.fetal_brain_male','fetal_brain_male','String','Cell type: fetal_brain_male'),
 ('f_0018','va.segway','va.segway.fetal_heart','fetal_heart','String','Cell type: fetal_heart'),
 ('f_0018','va.segway','va.segway.fetal_intestine_large','fetal_intestine_large','String','Cell type: fetal_intestine_large'),
 ('f_0018','va.segway','va.segway.fetal_intestine_small','fetal_intestine_small','String','Cell type: fetal_intestine_small'),
 ('f_0018','va.segway','va.segway.fetal_kidney','fetal_kidney','String','Cell type: fetal_kidney'),
 ('f_0018','va.segway','va.segway.fetal_lung','fetal_lung','String','Cell type: fetal_lung'),
 ('f_0018','va.segway','va.segway.fetal_muscle_leg','fetal_muscle_leg','String','Cell type: fetal_muscle_leg'),
 ('f_0018','va.segway','va.segway.fetal_muscle_trunk','fetal_muscle_trunk','String','Cell type: fetal_muscle_trunk'),
 ('f_0018','va.segway','va.segway.fetal_placenta','fetal_placenta','String','Cell type: fetal_placenta'),
 ('f_0018','va.segway','va.segway.fetal_stomach','fetal_stomach','String','Cell type: fetal_stomach'),
 ('f_0018','va.segway','va.segway.fetal_thymus','fetal_thymus','String','Cell type: fetal_thymus'),
 ('f_0018','va.segway','va.segway.gastric','gastric','String','Cell type: gastric'),
 ('f_0018','va.segway','va.segway.gliobla','gliobla','String','Cell type: gliobla'),
 ('f_0018','va.segway','va.segway.gm06990','gm06990','String','Cell type: gm06990'),
 ('f_0018','va.segway','va.segway.gm12864','gm12864','String','Cell type: gm12864'),
 ('f_0018','va.segway','va.segway.gm12865','gm12865','String','Cell type: gm12865'),
 ('f_0018','va.segway','va.segway.gm12891','gm12891','String','Cell type: gm12891'),
 ('f_0018','va.segway','va.segway.gm12892','gm12892','String','Cell type: gm12892'),
 ('f_0018','va.segway','va.segway.h1_bmp4_derived_mesendoderm_cultured_cells','h1_bmp4_derived_mesendoderm_cultured_cells','String','Cell type: h1_bmp4_derived_mesendoderm_cultured_cells'),
 ('f_0018','va.segway','va.segway.h1_bmp4_derived_trophoblast_cultured_cells','h1_bmp4_derived_trophoblast_cultured_cells','String','Cell type: h1_bmp4_derived_trophoblast_cultured_cells'),
 ('f_0018','va.segway','va.segway.h1_derived_mesenchymal_stem_cells','h1_derived_mesenchymal_stem_cells','String','Cell type: h1_derived_mesenchymal_stem_cells'),
 ('f_0018','va.segway','va.segway.h1_derived_neuronal_progenitor_cultured_cells','h1_derived_neuronal_progenitor_cultured_cells','String','Cell type: h1_derived_neuronal_progenitor_cultured_cells'),
 ('f_0018','va.segway','va.segway.h1_hesc','h1_hesc','String','Cell type: h1_hesc'),
 ('f_0018','va.segway','va.segway.h7_hesc','h7_hesc','String','Cell type: h7_hesc'),
 ('f_0018','va.segway','va.segway.h9_derived_neuronal_progenitor_cultured_cells','h9_derived_neuronal_progenitor_cultured_cells','String','Cell type: h9_derived_neuronal_progenitor_cultured_cells'),
 ('f_0018','va.segway','va.segway.h9_derived_neuron_cultured_cells','h9_derived_neuron_cultured_cells','String','Cell type: h9_derived_neuron_cultured_cells'),
 ('f_0018','va.segway','va.segway.h9es','h9es','String','Cell type: h9es'),
 ('f_0018','va.segway','va.segway.hac','hac','String','Cell type: hac'),
 ('f_0018','va.segway','va.segway.ha_sp','ha_sp','String','Cell type: ha_sp'),
 ('f_0018','va.segway','va.segway.hbmec','hbmec','String','Cell type: hbmec'),
 ('f_0018','va.segway','va.segway.hcfaa','hcfaa','String','Cell type: hcfaa'),
 ('f_0018','va.segway','va.segway.hcm','hcm','String','Cell type: hcm'),
 ('f_0018','va.segway','va.segway.hcpepic','hcpepic','String','Cell type: hcpepic'),
 ('f_0018','va.segway','va.segway.hct_116','hct_116','String','Cell type: hct_116'),
 ('f_0018','va.segway','va.segway.heepic','heepic','String','Cell type: heepic'),
 ('f_0018','va.segway','va.segway.hek293','hek293','String','Cell type: hek293'),
 ('f_0018','va.segway','va.segway.hepg2','hepg2','String','Cell type: hepg2'),
 ('f_0018','va.segway','va.segway.hesc_derived_cd184_endoderm_cultured_cells','hesc_derived_cd184_endoderm_cultured_cells','String','Cell type: hesc_derived_cd184+_endoderm_cultured_cells'),
 ('f_0018','va.segway','va.segway.hesc_derived_cd56_ectoderm_cultured_cells','hesc_derived_cd56_ectoderm_cultured_cells','String','Cell type: hesc_derived_cd56+_ectoderm_cultured_cells'),
 ('f_0018','va.segway','va.segway.hff','hff','String','Cell type: hff'),
 ('f_0018','va.segway','va.segway.hff_myc','hff_myc','String','Cell type: hff_myc'),
 ('f_0018','va.segway','va.segway.hl_60','hl_60','String','Cell type: hl_60'),
 ('f_0018','va.segway','va.segway.hmec','hmec','String','Cell type: hmec'),
 ('f_0018','va.segway','va.segway.hmf','hmf','String','Cell type: hmf'),
 ('f_0018','va.segway','va.segway.hpaf','hpaf','String','Cell type: hpaf'),
 ('f_0018','va.segway','va.segway.hpf','hpf','String','Cell type: hpf'),
 ('f_0018','va.segway','va.segway.hre','hre','String','Cell type: hre'),
 ('f_0018','va.segway','va.segway.hrpepic','hrpepic','String','Cell type: hrpepic'),
 ('f_0018','va.segway','va.segway.hsmm','hsmm','String','Cell type: hsmm'),
 ('f_0018','va.segway','va.segway.hsmmtube','hsmmtube','String','Cell type: hsmmtube'),
 ('f_0018','va.segway','va.segway.hues48_cell_line','hues48_cell_line','String','Cell type: hues48_cell_line'),
 ('f_0018','va.segway','va.segway.hues64_cell_line','hues64_cell_line','String','Cell type: hues64_cell_line'),
 ('f_0018','va.segway','va.segway.hues6_cell_line','hues6_cell_line','String','Cell type: hues6_cell_line'),
 ('f_0018','va.segway','va.segway.huvec','huvec','String','Cell type: huvec'),
 ('f_0018','va.segway','va.segway.imr90','imr90','String','Cell type: imr90'),
 ('f_0018','va.segway','va.segway.ips_15b_cell_line','ips_15b_cell_line','String','Cell type: ips_15b_cell_line'),
 ('f_0018','va.segway','va.segway.ips_18_cell_line','ips_18_cell_line','String','Cell type: ips_18_cell_line'),
 ('f_0018','va.segway','va.segway.ips_20b_cell_line','ips_20b_cell_line','String','Cell type: ips_20b_cell_line'),
 ('f_0018','va.segway','va.segway.ips_df_19_11_cell_line','ips_df_19_11_cell_line','String','Cell type: ips_df_19_11_cell_line'),
 ('f_0018','va.segway','va.segway.ips_df_6_9_cell_line','ips_df_6_9_cell_line','String','Cell type: ips_df_6_9_cell_line'),
 ('f_0018','va.segway','va.segway.k562','k562','String','Cell type: k562'),
 ('f_0018','va.segway','va.segway.left_ventricle','left_ventricle','String','Cell type: left_ventricle'),
 ('f_0018','va.segway','va.segway.lhcn_m2','lhcn_m2','String','Cell type: lhcn_m2'),
 ('f_0018','va.segway','va.segway.mcf_7','mcf_7','String','Cell type: mcf_7'),
 ('f_0018','va.segway','va.segway.mesenchymal_stem_cell_derived_adipocyte_cultured_cells','mesenchymal_stem_cell_derived_adipocyte_cultured_cells','String','Cell type: mesenchymal_stem_cell_derived_adipocyte_cultured_cells'),
 ('f_0018','va.segway','va.segway.mobilized_cd34_primary_cells_female','mobilized_cd34_primary_cells_female','String','Cell type: mobilized_cd34_primary_cells_female'),
 ('f_0018','va.segway','va.segway.mobilized_cd34_primary_cells_male','mobilized_cd34_primary_cells_male','String','Cell type: mobilized_cd34_primary_cells_male'),
 ('f_0018','va.segway','va.segway.monocytes_cd14_ro01746','monocytes_cd14_ro01746','String','Cell type: monocytes_cd14+_ro01746'),
 ('f_0018','va.segway','va.segway.nb4','nb4','String','Cell type: nb4'),
 ('f_0018','va.segway','va.segway.nh_a','nh_a','String','Cell type: nh_a'),
 ('f_0018','va.segway','va.segway.nhdf_ad','nhdf_ad','String','Cell type: nhdf_ad'),
 ('f_0018','va.segway','va.segway.nhdf_neo','nhdf_neo','String','Cell type: nhdf_neo'),
 ('f_0018','va.segway','va.segway.nhek','nhek','String','Cell type: nhek'),
 ('f_0018','va.segway','va.segway.nhlf','nhlf','String','Cell type: nhlf'),
 ('f_0018','va.segway','va.segway.nt2_d1','nt2_d1','String','Cell type: nt2_d1'),
 ('f_0018','va.segway','va.segway.osteobl','osteobl','String','Cell type: osteobl'),
 ('f_0018','va.segway','va.segway.ovary','ovary','String','Cell type: ovary'),
 ('f_0018','va.segway','va.segway.panc_1','panc_1','String','Cell type: panc_1'),
 ('f_0018','va.segway','va.segway.panislets','panislets','String','Cell type: panislets'),
 ('f_0018','va.segway','va.segway.penis_foreskin_fibroblast_primary_cells_skin01','penis_foreskin_fibroblast_primary_cells_skin01','String','Cell type: penis_foreskin_fibroblast_primary_cells_skin01'),
 ('f_0018','va.segway','va.segway.penis_foreskin_fibroblast_primary_cells_skin02','penis_foreskin_fibroblast_primary_cells_skin02','String','Cell type: penis_foreskin_fibroblast_primary_cells_skin02'),
 ('f_0018','va.segway','va.segway.penis_foreskin_keratinocyte_primary_cells_skin02','penis_foreskin_keratinocyte_primary_cells_skin02','String','Cell type: penis_foreskin_keratinocyte_primary_cells_skin02'),
 ('f_0018','va.segway','va.segway.penis_foreskin_keratinocyte_primary_cells_skin03','penis_foreskin_keratinocyte_primary_cells_skin03','String','Cell type: penis_foreskin_keratinocyte_primary_cells_skin03'),
 ('f_0018','va.segway','va.segway.penis_foreskin_melanocyte_primary_cells_skin01','penis_foreskin_melanocyte_primary_cells_skin01','String','Cell type: penis_foreskin_melanocyte_primary_cells_skin01'),
 ('f_0018','va.segway','va.segway.penis_foreskin_melanocyte_primary_cells_skin03','penis_foreskin_melanocyte_primary_cells_skin03','String','Cell type: penis_foreskin_melanocyte_primary_cells_skin03'),
 ('f_0018','va.segway','va.segway.peripheral_blood_mononuclear_primary_cells','peripheral_blood_mononuclear_primary_cells','String','Cell type: peripheral_blood_mononuclear_primary_cells'),
 ('f_0018','va.segway','va.segway.placenta_amnion','placenta_amnion','String','Cell type: placenta_amnion'),
 ('f_0018','va.segway','va.segway.progfib','progfib','String','Cell type: progfib'),
 ('f_0018','va.segway','va.segway.psoas_muscle_oc','psoas_muscle_oc','String','Cell type: psoas_muscle_oc'),
 ('f_0018','va.segway','va.segway.rectal_mucosa_donor_29','rectal_mucosa_donor_29','String','Cell type: rectal_mucosa_donor_29'),
 ('f_0018','va.segway','va.segway.rectal_mucosa_donor_31','rectal_mucosa_donor_31','String','Cell type: rectal_mucosa_donor_31'),
 ('f_0018','va.segway','va.segway.rectal_smooth_muscle','rectal_smooth_muscle','String','Cell type: rectal_smooth_muscle'),
 ('f_0018','va.segway','va.segway.right_atrium','right_atrium','String','Cell type: right_atrium'),
 ('f_0018','va.segway','va.segway.right_ventricle','right_ventricle','String','Cell type: right_ventricle'),
 ('f_0018','va.segway','va.segway.rptec','rptec','String','Cell type: rptec'),
 ('f_0018','va.segway','va.segway.saec','saec','String','Cell type: saec'),
 ('f_0018','va.segway','va.segway.sigmoid_colon','sigmoid_colon','String','Cell type: sigmoid_colon'),
 ('f_0018','va.segway','va.segway.skeletal_muscle_female','skeletal_muscle_female','String','Cell type: skeletal_muscle_female'),
 ('f_0018','va.segway','va.segway.skeletal_muscle_male','skeletal_muscle_male','String','Cell type: skeletal_muscle_male'),
 ('f_0018','va.segway','va.segway.sk_n_mc','sk_n_mc','String','Cell type: sk_n_mc'),
 ('f_0018','va.segway','va.segway.sk_n_sh','sk_n_sh','String','Cell type: sk_n_sh'),
 ('f_0018','va.segway','va.segway.sk_n_sh_ra','sk_n_sh_ra','String','Cell type: sk_n_sh_ra'),
 ('f_0018','va.segway','va.segway.small_intestine_oc','small_intestine_oc','String','Cell type: small_intestine_oc'),
 ('f_0018','va.segway','va.segway.stomach_mucosa','stomach_mucosa','String','Cell type: stomach_mucosa'),
 ('f_0018','va.segway','va.segway.stomach_smooth_muscle','stomach_smooth_muscle','String','Cell type: stomach_smooth_muscle'),
 ('f_0018','va.segway','va.segway.t_47d','t_47d','String','Cell type: t_47d'),
 ('f_0018','va.segway','va.segway.thymus','thymus','String','Cell type: thymus'),
 ('f_0018','va.segway','va.segway.weri_rb_1','weri_rb_1','String','Cell type: weri_rb_1'),
 ('f_0018','va.segway','va.segway.wi_38','wi_38','String','Cell type: wi_38'),
 ('f_0018','va.segway','va.segway.hesc_derived_cd56_mesoderm_cultured_cells','hesc_derived_cd56_mesoderm_cultured_cells','String','Cell type: hesc_derived_cd56+_mesoderm_cultured_cells'),
 ('vep','va.vep','va.vep.assembly_name','assembly_name','String',''),
 ('vep','va.vep','va.vep.allele_string','allele_string','String',''),
 ('vep','va.vep','va.vep.ancestral','ancestral','String',''),
 ('vep','va.vep','va.vep.colocated_variants','colocated_variants','Array[Struct]',''),
 ('vep','va.vep','va.vep.context','context','String',''),
 ('vep','va.vep','va.vep.end','end','Int',''),
 ('vep','va.vep','va.vep.id','id','String',''),
 ('vep','va.vep','va.vep.input','input','String',''),
 ('vep','va.vep','va.vep.intergenic_consequences','intergenic_consequences','Array[Struct]',''),
 ('vep','va.vep','va.vep.most_severe_consequence','most_severe_consequence','String',''),
 ('vep','va.vep','va.vep.motif_feature_consequences','motif_feature_consequences','Array[Struct]',''),
 ('vep','va.vep','va.vep.regulatory_feature_consequences','regulatory_feature_consequences','Array[Struct]',''),
 ('vep','va.vep','va.vep.seq_region_name','seq_region_name','String',''),
 ('vep','va.vep','va.vep.start','start','Int',''),
 ('vep','va.vep','va.vep.strand','strand','Int',''),
 ('vep','va.vep','va.vep.transcript_consequences','transcript_consequences','Array[Struct]',''),
 ('vep','va.vep','va.vep.variant_class','variant_class','String',''),
 ('f_0013','va.gnomAD.exomes','va.gnomAD.exomes.AN','AN','Int',NULL),
 ('f_0014','va.gnomAD.genomes','va.gnomAD.genomes.AN','AN','Int',NULL);
COMMIT;
